// moxygen/dist/parser.js
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { Parser as XmlParser } from 'xml2js';
import { createCompound, findCompound } from './compound.js';
import { inline, getAnchor } from './helpers.js';
import { log } from './logger.js';
import * as md from './markdown.js';
let parserOptions;
const references = {};
let root;
/**
 * Convert a Doxygen XML element tree to Markdown.
 */
function toMarkdown(element, context = []) {
    if (typeof element === 'string') {
        return element;
    }
    if (Array.isArray(element)) {
        return element.map((v) => toMarkdown(v, context)).join('');
    }
    if (typeof element !== 'object' || element === null) {
        return '';
    }
    const el = element;
    let s = '';
    // Opening
    switch (el['#name']) {
        case 'ref':
            return s + md.refLink(toMarkdown(el.$$), el.$?.refid ?? '');
        case '__text__':
            s = el._ ?? '';
            break;
        case 'emphasis':
            s = '*';
            break;
        case 'bold':
            s = '**';
            break;
        case 'parametername':
        case 'computeroutput':
            s = '`';
            break;
        case 'parameterlist':
            s = el.$?.kind === 'exception'
                ? '\n#### Exceptions\n'
                : '\n#### Parameters\n';
            break;
        case 'parameteritem':
            s = '* ';
            break;
        case 'programlisting':
            s = '\n```cpp\n';
            break;
        case 'orderedlist':
            context.push(el);
            s = '\n\n';
            break;
        case 'itemizedlist':
            s = '\n\n';
            break;
        case 'listitem':
            s = context.length > 0 && context[context.length - 1]['#name'] === 'orderedlist'
                ? '1. '
                : '* ';
            break;
        case 'sp':
            s = ' ';
            break;
        case 'heading':
            s = '## ';
            break;
        case 'xrefsect':
            s += '\n> ';
            break;
        case 'simplesect': {
            const kind = el.$?.kind;
            if (kind === 'attention' || kind === 'warning') {
                s = '\n:::warning\n';
            }
            else if (kind === 'note' || kind === 'remark') {
                s = '\n:::note\n';
            }
            else if (kind === 'deprecated') {
                s = '\n:::warning\n**Deprecated.** ';
            }
            else if (kind === 'return') {
                s = '\n#### Returns\n';
            }
            else if (kind === 'see') {
                s = '**See also**: ';
            }
            else if (kind === 'since') {
                s = '**Since**: ';
            }
            else {
                log.warn(`simplesect kind '${kind}' not supported`);
            }
            break;
        }
        case 'formula': {
            let formula = trim(el._ ?? '');
            if (formula.startsWith('$') && formula.endsWith('$'))
                return formula;
            if (formula.startsWith('\\[') && formula.endsWith('\\]')) {
                formula = trim(formula.substring(2, formula.length - 2));
            }
            return `\n$$\n${formula}\n$$\n`;
        }
        case 'preformatted':
            s = '\n<pre>';
            break;
        case 'sect1':
        case 'sect2':
        case 'sect3':
            context.push(el);
            s = `\n${getAnchor(el.$?.id ?? '', parserOptions)}\n`;
            break;
        case 'title': {
            const level = '#'.repeat(Number(context[context.length - 1]?.['#name']?.slice(-1) ?? '1'));
            s = `\n#${level} ${el._ ?? ''}\n`;
            break;
        }
        case 'mdash':
            s = '&mdash;';
            break;
        case 'ndash':
            s = '&ndash;';
            break;
        case 'linebreak':
            s = '<br/>';
            break;
        case 'xreftitle':
        case 'entry':
        case 'row':
        case 'ulink':
        case 'codeline':
        case 'highlight':
        case 'table':
        case 'para':
        case 'parameterdescription':
        case 'parameternamelist':
        case 'xrefdescription':
        case 'verbatim':
        case 'hruler':
        case undefined:
            break;
        default:
            log.warn(`${el['#name']}: not yet supported`);
    }
    // Recurse on children
    if (el.$$) {
        s += toMarkdown(el.$$, context);
    }
    // Closing
    switch (el['#name']) {
        case 'simplesect': {
            const closeKind = el.$?.kind;
            if (closeKind === 'attention' || closeKind === 'warning' || closeKind === 'note' || closeKind === 'remark' || closeKind === 'deprecated') {
                s += '\n:::\n\n';
            }
            else {
                s += '\n\n';
            }
            break;
        }
        case 'parameterlist':
        case 'para':
            s += '\n\n';
            break;
        case 'emphasis':
            s += '*';
            break;
        case 'bold':
            s += '**';
            break;
        case 'parameteritem':
            s += '\n';
            break;
        case 'computeroutput':
            s += '`';
            break;
        case 'parametername':
            s += '` ';
            break;
        case 'entry':
            s = md.escape.cell(s) + '|';
            break;
        case 'programlisting':
            s += '```\n';
            break;
        case 'codeline':
            s += '\n';
            break;
        case 'ulink':
            s = md.link(s, el.$?.url ?? '');
            break;
        case 'orderedlist':
            context.pop();
            s += '\n';
            break;
        case 'itemizedlist':
        case 'listitem':
            s += '\n';
            break;
        case 'xreftitle':
            s += ': ';
            break;
        case 'preformatted':
            s += '</pre>\n';
            break;
        case 'sect1':
        case 'sect2':
        case 'sect3':
            context.pop();
            s += '\n';
            break;
        case 'row':
            s = '\n' + md.escape.row(s);
            if (el.$$ && el.$$[0]?.$?.thead === 'yes') {
                s += el.$$.map((_, i) => (i ? ' | ' : '\n') + '---------').join('');
            }
            break;
    }
    return s;
}
function trim(text) {
    return text.trim();
}
/** Convert a named XML field to trimmed Markdown. */
function mdField(def, property) {
    return trim(toMarkdown(def[property]));
}
/** Extract a summary from brief or detailed description. */
function mdSummary(def) {
    let summary = trim(toMarkdown(def['briefdescription']));
    if (!summary) {
        summary = trim(toMarkdown(def['detaileddescription']));
        if (summary) {
            const firstSentence = summary.split('\n', 1)[0];
            if (firstSentence)
                summary = firstSentence;
        }
    }
    return summary;
}
function parseMembers(compound, props, membersdef) {
    Object.assign(compound, props);
    references[compound.refid] = compound;
    if (membersdef) {
        for (const memberdef of membersdef) {
            const member = {
                name: memberdef.name[0],
                parent: compound,
                refid: '',
                kind: '',
                section: '',
                prot: '',
                static: '',
                virtual: '',
                proto: '',
                briefdescription: '',
                detaileddescription: '',
                summary: '',
                enumvalue: [],
                returnType: '',
                params: [],
                templateParams: [],
                qualifiers: [],
                definition: '',
                argsstring: '',
                initializer: '',
                isConst: false,
                isInline: false,
                isExplicit: false,
                isStatic: false,
                isVirtual: false,
            };
            compound.members.push(member);
            const attrs = memberdef.$;
            if (attrs) {
                Object.assign(member, attrs);
            }
            references[member.refid] = member;
        }
    }
}
/** Extract parameter descriptions from detaileddescription XML. */
function extractParamDescriptions(memberdef) {
    const descriptions = {};
    const detailed = memberdef.detaileddescription;
    if (!detailed || !detailed.$$)
        return descriptions;
    const elements = detailed.$$;
    for (const para of elements) {
        if (para['#name'] !== 'para' || !para.$$)
            continue;
        for (const child of para.$$) {
            if (child['#name'] !== 'parameterlist')
                continue;
            if (!child.$$)
                continue;
            for (const item of child.$$) {
                if (item['#name'] !== 'parameteritem' || !item.$$)
                    continue;
                let paramName = '';
                let paramDesc = '';
                for (const part of item.$$) {
                    if (part['#name'] === 'parameternamelist' && part.$$) {
                        for (const pn of part.$$) {
                            if (pn['#name'] === 'parametername') {
                                paramName = toMarkdown(pn).trim();
                            }
                        }
                    }
                    else if (part['#name'] === 'parameterdescription') {
                        paramDesc = trim(toMarkdown(part));
                    }
                }
                if (paramName)
                    descriptions[paramName] = paramDesc;
            }
        }
    }
    return descriptions;
}
function parseMember(member, section, memberdef) {
    log.verbose(`Processing member ${member.kind} ${member.name}`);
    member.section = section;
    member.briefdescription = mdField(memberdef, 'briefdescription');
    member.detaileddescription = mdField(memberdef, 'detaileddescription');
    member.summary = mdSummary(memberdef);
    const attrs = memberdef.$;
    // Populate structured fields
    member.isConst = attrs['const'] === 'yes';
    member.isInline = attrs.inline === 'yes';
    member.isExplicit = attrs.explicit === 'yes';
    member.isStatic = attrs.static === 'yes';
    member.isVirtual = attrs.virt === 'virtual';
    member.returnType = trim(toMarkdown(memberdef.type));
    if (memberdef.location) {
        const locations = memberdef.location;
        const location = locations?.[0]?.$?.file;
        if (location) {
            member.location = location;
        }
    }
    // Definition and argsstring
    if (memberdef.definition) {
        const defArr = memberdef.definition;
        member.definition = defArr?.[0]?._ ?? '';
    }
    if (memberdef.argsstring) {
        const argsArr = memberdef.argsstring;
        member.argsstring = argsArr?.[0]?._ ?? '';
    }
    // Extract parameter descriptions from the detailed description XML
    const paramDescs = extractParamDescriptions(memberdef);
    // Template parameters
    if (memberdef.templateparamlist) {
        const tpl = memberdef.templateparamlist;
        if (tpl.length > 0 && tpl[0].param) {
            const params = tpl[0].param;
            for (const param of params) {
                member.templateParams.push({
                    type: trim(toMarkdown(param.type)),
                    name: param.declname ? trim(toMarkdown(param.declname)) : '',
                    description: '',
                });
            }
        }
    }
    // Function parameters
    if (memberdef.param && (member.kind === 'function' || member.kind === 'signal' || member.kind === 'slot')) {
        const params = memberdef.param;
        for (const param of params) {
            const paramName = param.declname ? trim(toMarkdown(param.declname)) : '';
            member.params.push({
                type: trim(toMarkdown(param.type)),
                name: paramName,
                description: paramDescs[paramName] ?? '',
            });
        }
    }
    // Qualifiers
    const qualifiers = [];
    if (member.isConst)
        qualifiers.push('const');
    if (member.argsstring) {
        if (/noexcept$/.test(member.argsstring))
            qualifiers.push('noexcept');
        if (/=\s*delete$/.test(member.argsstring))
            qualifiers.push('= delete');
        if (/=\s*default/.test(member.argsstring))
            qualifiers.push('= default');
    }
    member.qualifiers = qualifiers;
    // Build legacy proto string (for classic templates)
    let m = [];
    switch (member.kind) {
        case 'signal':
        case 'slot':
            m.push(`{${member.kind}} `);
        // fallthrough
        case 'function':
            m.push(attrs.prot, ' ');
            if (member.templateParams.length > 0) {
                m.push('template<');
                member.templateParams.forEach((tp, i) => {
                    if (i > 0)
                        m.push(',');
                    m.push(tp.type);
                    if (tp.name)
                        m.push(' ', tp.name);
                });
                m.push('>  \n');
            }
            if (member.isInline)
                m.push('inline', ' ');
            if (member.isStatic)
                m.push('static', ' ');
            if (member.isVirtual)
                m.push('virtual', ' ');
            m.push(toMarkdown(memberdef.type), ' ');
            if (member.isExplicit)
                m.push('explicit', ' ');
            m.push(md.refLink(member.name, member.refid));
            m.push('(');
            member.params.forEach((param, i) => {
                if (i > 0)
                    m.push(', ');
                m.push(param.type);
                if (param.name)
                    m.push(' ', param.name);
            });
            m.push(')');
            for (const q of qualifiers) {
                m.push(' ', q);
            }
            break;
        case 'variable':
            m.push(attrs.prot, ' ');
            if (member.isStatic)
                m.push('static', ' ');
            if (attrs.mutable === 'yes')
                m.push('mutable', ' ');
            m.push(toMarkdown(memberdef.type), ' ');
            m.push(md.refLink(member.name, member.refid));
            if (memberdef.initializer) {
                const init = memberdef.initializer[0]?._ ?? '';
                if (init) {
                    m.push(' ', init);
                    member.initializer = init;
                }
            }
            break;
        case 'property':
            m.push(`{${member.kind}} `);
            m.push(toMarkdown(memberdef.type), ' ');
            m.push(md.refLink(member.name, member.refid));
            break;
        case 'enum':
            member.enumvalue = [];
            if (memberdef.enumvalue) {
                for (const param of memberdef.enumvalue) {
                    member.enumvalue.push({
                        name: mdField(param, 'name'),
                        briefdescription: mdField(param, 'briefdescription'),
                        detaileddescription: mdField(param, 'detaileddescription'),
                        summary: mdSummary(param),
                    });
                }
            }
            m.push(member.kind, ' ', md.refLink(member.name, member.refid));
            break;
        default:
            m.push(member.kind, ' ', md.refLink(member.name, member.refid));
            break;
    }
    member.proto = inline(m);
}
function assignToNamespace(compound, child) {
    if (compound.name !== child.namespace) {
        log.warn(`namespace mismatch: ${compound.name} != ${child.namespace}`);
    }
    if (child.parent) {
        delete child.parent.compounds[child.id];
    }
    compound.compounds[child.id] = child;
    child.parent = compound;
}
function assignCompoundGroup(compound, group) {
    compound.groupid = group.id;
    compound.groupname = group.name;
    for (const member of compound.members) {
        member.groupid = group.id;
        member.groupname = group.name;
    }
    for (const child of Object.values(compound.compounds)) {
        assignCompoundGroup(child, group);
    }
}
function pruneGroupTopLevelDuplicates(group) {
    const topLevelRefids = new Set(Object.values(group.compounds).map((compound) => compound.refid));
    for (const [id, compound] of Object.entries(group.compounds)) {
        let current = compound.parent;
        while (current) {
            if (topLevelRefids.has(current.refid) && current.groupid === group.id) {
                delete group.compounds[id];
                break;
            }
            current = current.parent;
        }
    }
}
function assignNamespaceToGroup(compound, child) {
    compound.compounds[child.id] = child;
    for (const id of Object.keys(child.compounds)) {
        delete compound.compounds[id];
    }
}
function assignClassToGroup(compound, child) {
    compound.compounds[child.id] = child;
    assignCompoundGroup(child, compound);
}
function extractPageSections(page, elements) {
    for (const element of elements) {
        if (element['#name'] === 'sect1' || element['#name'] === 'sect2' || element['#name'] === 'sect3') {
            const id = element.$?.id ?? '';
            const member = {
                section: element['#name'],
                id,
                name: id,
                refid: id,
                parent: page,
                kind: '',
                prot: '',
                static: '',
                virtual: '',
                proto: '',
                briefdescription: '',
                detaileddescription: '',
                summary: '',
                enumvalue: [],
                returnType: '',
                params: [],
                templateParams: [],
                qualifiers: [],
                definition: '',
                argsstring: '',
                initializer: '',
                isConst: false,
                isInline: false,
                isExplicit: false,
                isStatic: false,
                isVirtual: false,
            };
            page.members.push(member);
            references[member.refid] = member;
        }
        if (element.$$) {
            extractPageSections(page, element.$$);
        }
    }
}
function parseCompound(compound, compounddef) {
    log.verbose(`Processing compound ${compound.name}`);
    const attrs = compounddef.$;
    Object.assign(compound, attrs);
    compound.fullname = compounddef.compoundname[0]._ ??
        compounddef.compoundname[0] ?? '';
    // Short name: strip namespace prefix
    const nameParts = compound.fullname.split('::');
    compound.shortname = nameParts[nameParts.length - 1] || compound.name;
    // Group title (from @defgroup "Title")
    if (compounddef.title) {
        const titleArr = compounddef.title;
        const title = titleArr?.[0]?._ ?? titleArr?.[0] ?? '';
        if (title)
            compound.shortname = title;
    }
    // Language from XML attribute
    if (attrs.language) {
        compound.language = attrs.language;
    }
    // Include file
    if (compounddef.includes) {
        const includes = compounddef.includes;
        if (includes.length > 0) {
            compound.includes = includes[0]._ ?? '';
        }
    }
    compound.briefdescription = mdField(compounddef, 'briefdescription');
    compound.detaileddescription = mdField(compounddef, 'detaileddescription');
    compound.summary = mdSummary(compounddef);
    if (compounddef.basecompoundref) {
        for (const ref of compounddef.basecompoundref) {
            const refAttrs = ref.$;
            compound.basecompoundref.push({
                prot: refAttrs.prot,
                name: ref._,
                refid: refAttrs.refid,
            });
        }
    }
    if (compounddef.derivedcompoundref) {
        for (const ref of compounddef.derivedcompoundref) {
            const refAttrs = ref.$;
            compound.derivedcompoundref.push({
                prot: refAttrs.prot,
                name: ref._,
                refid: refAttrs.refid,
            });
        }
    }
    if (compounddef.sectiondef) {
        for (const section of compounddef.sectiondef) {
            const sectionAttrs = section.$;
            if (section.memberdef) {
                for (const memberdef of section.memberdef) {
                    const memberAttrs = memberdef.$;
                    const member = references[memberAttrs.id];
                    if (!member)
                        continue;
                    if (compound.kind === 'group') {
                        member.groupid = compound.id;
                        member.groupname = compound.name;
                    }
                    else if (compound.kind === 'file') {
                        root.members.push(member);
                    }
                    parseMember(member, sectionAttrs.kind, memberdef);
                }
            }
        }
    }
    compound.proto = inline([compound.kind, ' ', md.refLink(compound.name, compound.refid)]);
    if (compounddef.location) {
        const locations = compounddef.location;
        const location = locations?.[0]?.$?.file;
        if (location) {
            compound.location = location;
        }
    }
    // Handle innerclass for any compound that can contain nested types
    if (compounddef.innerclass) {
        const innerclassRefs = [];
        for (const innerclassdef of compounddef.innerclass) {
            const innerAttrs = innerclassdef.$;
            innerclassRefs.push(innerAttrs.refid);
            const ref = references[innerAttrs.refid];
            if (!ref)
                continue;
            switch (compound.kind) {
                case 'namespace':
                    assignToNamespace(compound, ref);
                    break;
                case 'group':
                    assignClassToGroup(compound, ref);
                    break;
                case 'file':
                    break;
                default:
                    // class, interface, enum - nest inner classes directly
                    compound.compounds[ref.id] = ref;
                    ref.parent = compound;
                    break;
            }
        }
        if (compound.kind === 'file') {
            compound.fileCompoundRefs = innerclassRefs;
        }
    }
    if (compounddef.innernamespace && compound.kind === 'file') {
        const innernamespaceRefs = [];
        for (const namespacedef of compounddef.innernamespace) {
            const nsAttrs = namespacedef.$;
            innernamespaceRefs.push(nsAttrs.refid);
        }
        compound.fileNamespaceRefs = innernamespaceRefs;
    }
    switch (compound.kind) {
        case 'class':
        case 'struct':
        case 'union':
        case 'typedef':
        case 'interface':
        case 'enum': {
            const parts = compound.name.split('::');
            compound.namespace = parts.slice(0, -1).join('::');
            break;
        }
        case 'page':
            extractPageSections(compound, compounddef.$$);
            break;
        case 'namespace':
        case 'group':
            if (compound.kind === 'group') {
                compound.groupid = compound.id;
                compound.groupname = compound.name;
            }
            if (compounddef.innernamespace) {
                compound.innernamespaces = [];
                for (const namespacedef of compounddef.innernamespace) {
                    const nsAttrs = namespacedef.$;
                    if (compound.kind === 'group') {
                        const ref = references[nsAttrs.refid];
                        if (ref)
                            assignNamespaceToGroup(compound, ref);
                    }
                }
            }
            if (compound.kind === 'group') {
                pruneGroupTopLevelDuplicates(compound);
            }
            break;
        default:
            break;
    }
}
function parseIndex(rootCompound, index, options) {
    // Pass 1: register all compounds and their members in the references map.
    // Sort by refid length (ascending) so parent namespaces are created before children.
    const sorted = [...index].sort((a, b) => {
        const aRefid = a.$.refid ?? '';
        const bRefid = b.$.refid ?? '';
        return aRefid.length - bRefid.length;
    });
    for (const element of sorted) {
        const attrs = element.$;
        const name = element.name?.[0] ?? '';
        const compound = findCompound(rootCompound, attrs.refid, name, true);
        parseMembers(compound, attrs, element.member);
    }
    // Pass 2: parse compound XML files. All refs are now registered so
    // innerclass/innernamespace lookups succeed regardless of order.
    for (const element of sorted) {
        const attrs = element.$;
        const compound = references[attrs.refid];
        if (!compound)
            continue;
        const xmlPath = join(options.directory, `${compound.refid}.xml`);
        log.verbose(`Parsing ${xmlPath}`);
        try {
            const xml = readFileSync(xmlPath, 'utf8');
            const xmlParser = new XmlParser({
                explicitChildren: true,
                preserveChildrenOrder: true,
                charsAsChildren: true,
            });
            xmlParser.parseString(xml, (err, data) => {
                if (err) {
                    log.warn(`parse error for file: ${xmlPath}`);
                    return;
                }
                const doxygen = data.doxygen;
                const compounddefs = doxygen.compounddef;
                parseCompound(compound, compounddefs[0]);
            });
        }
        catch (e) {
            log.warn(`failed to read: ${xmlPath}`);
        }
    }
}
/**
 * Load and parse Doxygen XML index.
 */
export async function loadIndex(options) {
    parserOptions = options;
    root = createCompound();
    root.kind = 'index';
    // Clear references
    for (const key of Object.keys(references)) {
        delete references[key];
    }
    const indexPath = join(options.directory, 'index.xml');
    log.verbose(`Parsing ${indexPath}`);
    const xml = readFileSync(indexPath, 'utf8');
    return new Promise((resolve, reject) => {
        const xmlParser = new XmlParser();
        xmlParser.parseString(xml, (err, result) => {
            if (err) {
                reject(new Error(`Failed to parse doxygen XML: ${err}`));
                return;
            }
            const doxygenindex = result.doxygenindex;
            const compounds = doxygenindex.compound;
            parseIndex(root, compounds, options);
            resolve({ root, references });
        });
    });
}
