// sourcey/dist/config.js
import { access, readdir } from "node:fs/promises";
import { resolve, relative, dirname, basename, extname } from "node:path";
import { createJiti } from "jiti";
export function defineConfig(config) {
    return config;
}
// ---------------------------------------------------------------------------
// Config loading
// ---------------------------------------------------------------------------
const DEFAULT_COLORS = {
    primary: "99 102 241",
    light: "129 140 248",
    dark: "79 70 229",
};
const DEFAULT_FONT_SANS = "Inter";
const SYSTEM_SANS = "ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif";
const SYSTEM_MONO = "ui-monospace, 'SF Mono', 'Cascadia Code', Consolas, 'Liberation Mono', Menlo, monospace";
const DEFAULT_LAYOUT = {
    sidebar: "18rem",
    toc: "19rem",
    content: "44rem",
};
export async function loadConfig(cwd) {
    const dir = cwd ?? process.cwd();
    const configPath = resolve(dir, "sourcey.config.ts");
    try {
        await access(configPath);
    }
    catch {
        throw new Error(`No sourcey.config.ts found in ${dir}.\n` +
            `Create one with:\n\n` +
            `  import { defineConfig } from "sourcey";\n` +
            `  export default defineConfig({ navigation: { tabs: [...] } });\n`);
    }
    const jiti = createJiti(import.meta.url);
    const mod = await jiti.import(configPath);
    const raw = mod.default;
    if (!raw?.navigation?.tabs) {
        throw new Error("sourcey.config.ts must export default defineConfig({ navigation: { tabs: [...] } })");
    }
    return resolveConfigFromRaw(raw, dirname(configPath));
}
/**
 * Build a ResolvedConfig for a standalone spec file (no config file needed).
 */
export function configFromSpec(specPath) {
    const absSpec = resolve(specPath);
    return {
        name: "API Reference",
        theme: {
            preset: "default",
            colors: { ...DEFAULT_COLORS },
            fonts: { sans: `'${DEFAULT_FONT_SANS}', ${SYSTEM_SANS}`, mono: SYSTEM_MONO, googleFont: DEFAULT_FONT_SANS },
            layout: { ...DEFAULT_LAYOUT },
            css: [],
        },
        codeSamples: ["curl", "javascript", "python"],
        tabs: [{ label: "API Reference", slug: "api", openapi: absSpec }],
        navbar: { links: [] },
        footer: { links: [] },
        search: { featured: [] },
    };
}
// ---------------------------------------------------------------------------
// Resolution + validation
// ---------------------------------------------------------------------------
export async function resolveConfigFromRaw(raw, configDir) {
    const theme = resolveTheme(raw, configDir);
    const logo = resolveLogo(raw.logo, configDir);
    const tabs = await resolveTabs(raw.navigation.tabs, configDir);
    return {
        name: raw.name ?? "",
        theme,
        logo,
        favicon: raw.favicon && !raw.favicon.startsWith("http") && !raw.favicon.startsWith("data:") ? resolve(configDir, raw.favicon) : raw.favicon,
        repo: raw.repo,
        editBranch: raw.editBranch,
        editBasePath: raw.editBasePath,
        codeSamples: raw.codeSamples ?? ["curl", "javascript", "python"],
        tabs,
        navbar: {
            links: raw.navbar?.links ?? [],
            primary: raw.navbar?.primary,
        },
        footer: {
            links: raw.footer?.links ?? [],
        },
        search: {
            featured: raw.search?.featured ?? [],
        },
    };
}
function isUrl(source) {
    return source.startsWith("http://") || source.startsWith("https://");
}
const VALID_PRESETS = ["default", "minimal", "api-first"];
function resolveTheme(raw, configDir) {
    const preset = raw.theme?.preset ?? "default";
    if (!VALID_PRESETS.includes(preset)) {
        throw new Error(`Invalid theme preset "${preset}". Must be one of: ${VALID_PRESETS.join(", ")}`);
    }
    const rawColors = raw.theme?.colors;
    const colors = rawColors
        ? {
            primary: hexToRgb(rawColors.primary),
            light: rawColors.light ? hexToRgb(rawColors.light) : hexToRgb(rawColors.primary),
            dark: rawColors.dark ? hexToRgb(rawColors.dark) : hexToRgb(rawColors.primary),
        }
        : { ...DEFAULT_COLORS };
    const sansName = raw.theme?.fonts?.sans ?? DEFAULT_FONT_SANS;
    const monoName = raw.theme?.fonts?.mono;
    const fonts = {
        sans: `'${sansName}', ${SYSTEM_SANS}`,
        mono: monoName ? `'${monoName}', ${SYSTEM_MONO}` : SYSTEM_MONO,
        googleFont: sansName,
    };
    const layout = {
        sidebar: raw.theme?.layout?.sidebar ?? DEFAULT_LAYOUT.sidebar,
        toc: raw.theme?.layout?.toc ?? DEFAULT_LAYOUT.toc,
        content: raw.theme?.layout?.content ?? DEFAULT_LAYOUT.content,
    };
    const css = (raw.theme?.css ?? []).map(p => resolve(configDir, p));
    return { preset, colors, fonts, layout, css };
}
function resolveLogo(logo, configDir) {
    if (!logo)
        return undefined;
    const resolvePath = (p) => p && !p.startsWith("http") && !p.startsWith("data:") ? resolve(configDir, p) : p;
    if (typeof logo === "string")
        return { light: resolvePath(logo) };
    return { light: resolvePath(logo.light), dark: resolvePath(logo.dark), href: logo.href };
}
async function resolveTabs(tabs, configDir) {
    const slugs = new Set();
    const resolved = [];
    for (const tab of tabs) {
        if (!tab.tab)
            throw new Error("Tab missing \"tab\" name");
        const slug = tab.slug !== undefined ? tab.slug : slugify(tab.tab);
        if (slugs.has(slug))
            throw new Error(`Duplicate tab slug "${slug}" (from "${tab.tab}")`);
        slugs.add(slug);
        const sources = [tab.openapi, tab.groups, tab.doxygen, tab.mcp].filter(Boolean).length;
        if (sources > 1) {
            throw new Error(`Tab "${tab.tab}" has multiple sources; use only one of "openapi", "groups", "doxygen", or "mcp"`);
        }
        if (sources === 0) {
            throw new Error(`Tab "${tab.tab}" needs one of "openapi", "groups", "doxygen", or "mcp"`);
        }
        if (tab.openapi) {
            if (isUrl(tab.openapi)) {
                resolved.push({ label: tab.tab, slug, openapi: tab.openapi });
            }
            else {
                const absPath = resolve(configDir, tab.openapi);
                await assertExists(absPath, `OpenAPI spec "${tab.openapi}" in tab "${tab.tab}"`);
                resolved.push({ label: tab.tab, slug, openapi: absPath });
            }
        }
        else if (tab.mcp) {
            if (isUrl(tab.mcp)) {
                resolved.push({ label: tab.tab, slug, mcp: tab.mcp });
            }
            else {
                const absPath = resolve(configDir, tab.mcp);
                await assertExists(absPath, `MCP spec "${tab.mcp}" in tab "${tab.tab}"`);
                resolved.push({ label: tab.tab, slug, mcp: absPath });
            }
        }
        else if (tab.doxygen) {
            const absXml = resolve(configDir, tab.doxygen.xml);
            await assertExists(absXml, `Doxygen XML directory "${tab.doxygen.xml}" in tab "${tab.tab}"`);
            resolved.push({
                label: tab.tab,
                slug,
                doxygen: {
                    xml: absXml,
                    language: tab.doxygen.language ?? "cpp",
                    groups: tab.doxygen.groups ?? false,
                    index: tab.doxygen.index === false ? "none" : (tab.doxygen.index ?? "auto"),
                },
            });
        }
        else {
            const groups = await resolveGroups(tab.groups, tab.tab, configDir);
            resolved.push({ label: tab.tab, slug, groups });
        }
    }
    return resolved;
}
async function resolveGroups(groups, tabName, configDir) {
    const resolved = [];
    for (const group of groups) {
        if (!group.group)
            throw new Error(`Group missing "group" name in tab "${tabName}"`);
        if (!group.pages?.length)
            throw new Error(`Group "${group.group}" in tab "${tabName}" has no pages`);
        const pages = [];
        for (const pageSlug of group.pages) {
            if (pageSlug.includes("*")) {
                const expanded = await expandGlob(pageSlug, configDir);
                for (const file of expanded) {
                    const rel = relative(configDir, file).replace(/\.[^.]+$/, "");
                    pages.push({ slug: rel, file });
                }
            }
            else {
                const absPath = await resolvePagePath(pageSlug, configDir);
                pages.push({ slug: pageSlug, file: absPath });
            }
        }
        resolved.push({ label: group.group, pages });
    }
    return resolved;
}
/**
 * Expand a simple glob pattern like "doc/api-*" into matching .md/.mdx files.
 * Supports trailing * only (e.g. "doc/api-*", "guides/*").
 */
async function expandGlob(pattern, configDir) {
    const absPattern = resolve(configDir, pattern);
    const dir = dirname(absPattern);
    const prefix = basename(absPattern).replace("*", "");
    try {
        const entries = await readdir(dir);
        const matches = entries
            .filter((f) => {
            const ext = extname(f);
            if (ext !== ".md" && ext !== ".mdx")
                return false;
            const name = basename(f, ext);
            return prefix ? name.startsWith(prefix) : true;
        })
            .sort()
            .map((f) => resolve(dir, f));
        if (!matches.length) {
            throw new Error(`Glob "${pattern}" matched no files in ${dir}`);
        }
        return matches;
    }
    catch (err) {
        if (err.code === "ENOENT") {
            throw new Error(`Glob directory not found: ${dir}`);
        }
        throw err;
    }
}
async function resolvePagePath(slug, configDir) {
    for (const ext of [".md", ".mdx"]) {
        const candidate = resolve(configDir, `${slug}${ext}`);
        try {
            await access(candidate);
            return candidate;
        }
        catch { /* try next */ }
    }
    throw new Error(`Page "${slug}" not found (tried ${slug}.md, ${slug}.mdx in ${configDir})`);
}
// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------
export function slugify(name) {
    return name.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
}
/** Build a path under a tab, handling empty slugs (root-level tabs). */
export function tabPath(tabSlug, file) {
    if (!tabSlug)
        return file;
    if (file.startsWith(`${tabSlug}/`))
        return file;
    return `${tabSlug}/${file}`;
}
export function hexToRgb(hex) {
    const h = hex.replace("#", "");
    if (h.length !== 6)
        throw new Error(`Invalid hex color: "${hex}"`);
    const r = parseInt(h.slice(0, 2), 16);
    const g = parseInt(h.slice(2, 4), 16);
    const b = parseInt(h.slice(4, 6), 16);
    return `${r} ${g} ${b}`;
}
async function assertExists(filePath, label) {
    try {
        await access(filePath);
    }
    catch {
        throw new Error(`${label} not found: ${filePath}`);
    }
}
