// sourcey/dist/components/mcp/McpReturns.js
import { jsx as _jsx, jsxs as _jsxs } from "preact/jsx-runtime";
import { SchemaView } from "../schema/SchemaView.js";
import { ExampleView } from "../schema/ExampleView.js";
import { SectionLabel } from "../ui/SectionLabel.js";
/**
 * Left column: "Returns" section label + schema view or fallback text.
 */
export function McpReturnsCopy({ schema }) {
    return (_jsxs("div", { class: "mt-6", children: [_jsx(SectionLabel, { children: "Returns" }), schema ? (_jsx(SchemaView, { schema: schema })) : (_jsx("p", { class: "text-sm text-[rgb(var(--color-gray-500))]", children: "Returns MCP content array (text, image, or embedded resource)." }))] }));
}
/**
 * Right column (sticky panel): example JSON for the output schema.
 */
export function McpReturnsExample({ schema }) {
    if (!schema) {
        return (_jsx(ExampleView, { schema: {
                type: "object",
                properties: {
                    content: {
                        type: "array",
                        items: {
                            type: "object",
                            properties: {
                                type: { type: "string", enum: ["text", "image", "resource"] },
                                text: { type: "string", example: "..." },
                            },
                        },
                    },
                },
            }, title: "Response" }));
    }
    return _jsx(ExampleView, { schema: schema, title: "Response" });
}
