// sourcey/dist/core/doxygen-loader.js
import { generate } from "moxygen";
import { renderMarkdown, extractHeadings, extractFirstParagraph, stripMarkdownLinks } from "../utils/markdown.js";
export function normalizeDoxygenDescription(description, markdown) {
    if (!description)
        return "";
    if (!description.includes("{#ref "))
        return description;
    const firstParagraph = extractFirstParagraph(markdown);
    return firstParagraph || description;
}
export function rewriteGeneratedDoxygenIncludePath(includePath) {
    const normalized = includePath.replace(/\\/g, "/");
    if (!normalized.startsWith("/") && !/^[A-Za-z]:\//.test(normalized)) {
        return includePath;
    }
    const includeMarker = "/include/";
    const includeIndex = normalized.lastIndexOf(includeMarker);
    if (includeIndex !== -1) {
        return normalized.slice(includeIndex + includeMarker.length);
    }
    const srcMarker = "/src/";
    const srcIndex = normalized.indexOf(srcMarker);
    if (srcIndex !== -1) {
        return normalized.slice(srcIndex + 1);
    }
    const basename = normalized.split("/").pop();
    return basename || includePath;
}
export function rewriteGeneratedDoxygenMarkdown(markdown) {
    return markdown.replace(/^#include ([<"])([^>\n"]+)([>"])$/gm, (_match, open, includePath, close) => {
        return `#include ${open}${rewriteGeneratedDoxygenIncludePath(includePath)}${close}`;
    });
}
export function rewriteGeneratedDoxygenHref(href) {
    const [path, hash] = href.split("#", 2);
    if (!path.startsWith("api_") || !path.endsWith(".md"))
        return href;
    const slug = path
        .slice("api_".length, -".md".length)
        .replace(/--/g, "-");
    return `${slug}.html${hash ? `#${hash}` : ""}`;
}
export function rewriteGeneratedDoxygenHtmlLinks(html) {
    return html
        .replace(/href="(api_[^"]+?\.md(?:#[^"]*)?)"/g, (_match, href) => {
        return `href="${rewriteGeneratedDoxygenHref(href)}"`;
    })
        .replace(/<code>\[([^\]]+)\]\((api_[^)]+?\.md(?:#[^)]+)?)\)<\/code>/g, (_match, label, href) => {
        return `<a href="${rewriteGeneratedDoxygenHref(href)}"><code>${label}</code></a>`;
    });
}
export async function loadDoxygenTab(config, tabSlug, tabLabel) {
    const generated = await generate({
        directory: config.xml,
        language: config.language,
        quiet: true,
    });
    const pages = new Map();
    const groupMap = new Map();
    for (const page of generated) {
        if (!page.markdown.trim())
            continue;
        const markdown = rewriteGeneratedDoxygenMarkdown(page.markdown);
        const description = normalizeDoxygenDescription(page.description, markdown);
        const html = rewriteGeneratedDoxygenHtmlLinks(renderMarkdown(markdown));
        const headings = extractHeadings(markdown);
        pages.set(page.slug, {
            title: page.title,
            description,
            slug: page.slug,
            html,
            headings,
            sourcePath: `api/${page.slug}.md`,
            editPath: page.kind === "group" ? `api/${page.slug}.md` : null,
        });
        // Group by the last segment of namespace (e.g. "icy::http" -> "http")
        const groupKey = page.module ?? lastSegment(page.namespace) ?? "API";
        if (!groupMap.has(groupKey)) {
            groupMap.set(groupKey, []);
        }
        groupMap.get(groupKey).push({ slug: page.slug, title: page.title, kind: page.kind });
    }
    // Build navigation: namespace index pages first, then classes alphabetically
    const groups = [];
    for (const [key, items] of groupMap) {
        const groupPage = items.find((i) => i.kind === "group");
        const label = groupPage?.title ?? key;
        const sorted = items.sort((a, b) => {
            if (a.kind === "group" && b.kind !== "group")
                return -1;
            if (a.kind !== "group" && b.kind === "group")
                return 1;
            if (a.kind === "namespace" && b.kind !== "namespace")
                return -1;
            if (a.kind !== "namespace" && b.kind === "namespace")
                return 1;
            return a.title.localeCompare(b.title);
        });
        groups.push({
            label,
            items: sorted.map((p) => ({
                label: p.kind === "group" ? "Overview" : p.title,
                href: `${tabSlug}/${p.slug}.html`,
                id: p.slug,
            })),
        });
    }
    // Generate index page
    const indexStyle = resolveIndexStyle(config.index, groupMap, pages);
    if (indexStyle !== "none") {
        const indexSlug = "index";
        const indexHtml = buildIndexHtml(indexStyle, tabSlug, tabLabel, groupMap, pages);
        const indexHeadings = indexStyle === "rich" ? [] : extractHeadings(indexHtml);
        pages.set(indexSlug, {
            title: tabLabel,
            description: "",
            slug: indexSlug,
            html: indexHtml,
            headings: indexHeadings,
            sourcePath: `api/${indexSlug}.md`,
            editPath: null,
        });
        groups.unshift({
            label: tabLabel,
            items: [{ label: "Overview", href: `${tabSlug}/${indexSlug}.html`, id: indexSlug }],
        });
    }
    const firstItem = groups[0]?.items[0];
    return {
        pages,
        navTab: {
            label: tabLabel,
            slug: tabSlug,
            href: firstItem?.href ?? `${tabSlug}/`,
            kind: "docs",
            groups,
        },
    };
}
/**
 * Resolve "auto" to a concrete style based on the data we have.
 */
function resolveIndexStyle(configured, groupMap, pages) {
    if (configured !== "auto")
        return configured;
    // Tiny projects — no index needed
    if (pages.size < 5)
        return "none";
    // Check if we have groups with descriptions
    const groupsWithDesc = [...groupMap.values()]
        .map((items) => items.find((i) => i.kind === "group"))
        .filter((g) => !!g)
        .filter((g) => pages.get(g.slug)?.description);
    if (groupsWithDesc.length >= 2)
        return "rich";
    // Check if we have meaningful grouping (multiple groups/namespaces)
    if (groupMap.size >= 2)
        return "structured";
    return "flat";
}
/**
 * Build index page HTML using sourcey's existing component markup.
 */
function buildIndexHtml(style, _tabSlug, _tabLabel, groupMap, pages) {
    if (style === "rich")
        return buildRichIndex(groupMap, pages);
    if (style === "structured")
        return buildStructuredIndex(groupMap, pages);
    return buildFlatIndex(groupMap);
}
/**
 * Rich: card grid with module name, description, and type count.
 * Uses sourcey's card-group/card-item CSS.
 */
function buildRichIndex(groupMap, pages) {
    const cards = [];
    for (const [, items] of groupMap) {
        const groupEntry = items.find((i) => i.kind === "group");
        if (!groupEntry)
            continue;
        const page = pages.get(groupEntry.slug);
        if (!page)
            continue;
        const typeCount = items.filter((i) => i.kind !== "group" && i.kind !== "namespace").length;
        const href = `${groupEntry.slug}.html`;
        const desc = stripMarkdownLinks(page.description || "");
        const meta = typeCount > 0
            ? `<p style="margin:0.5rem 0 0;font-size:0.8rem;opacity:0.5">${typeCount} type${typeCount !== 1 ? "s" : ""}</p>`
            : "";
        cards.push(`<a href="${href}" class="card-item">` +
            `<div class="card-item-inner">` +
            `<h3 class="card-item-title">${escHtml(page.title)}</h3>` +
            (desc ? `<div class="card-item-content"><p>${escHtml(desc)}</p>${meta}</div>` : `<div class="card-item-content">${meta}</div>`) +
            `</div></a>`);
    }
    // Also include groups without a group page (orphan namespaces)
    for (const [key, items] of groupMap) {
        const hasGroup = items.some((i) => i.kind === "group");
        if (hasGroup)
            continue;
        const typeCount = items.filter((i) => i.kind !== "namespace").length;
        if (typeCount === 0)
            continue;
        const firstItem = items[0];
        const href = `${firstItem.slug}.html`;
        cards.push(`<a href="${href}" class="card-item">` +
            `<div class="card-item-inner">` +
            `<h3 class="card-item-title">${escHtml(key)}</h3>` +
            `<div class="card-item-content"><p style="margin:0;font-size:0.8rem;opacity:0.5">${typeCount} type${typeCount !== 1 ? "s" : ""}</p></div>` +
            `</div></a>`);
    }
    if (cards.length === 0)
        return "";
    const cols = cards.length <= 2 ? "2" : "3";
    return `<div class="card-group not-prose" data-cols="${cols}">\n${cards.join("\n")}\n</div>`;
}
/**
 * Structured: grouped list of types by module/namespace.
 * Rendered as markdown headings with link lists.
 */
function buildStructuredIndex(groupMap, pages) {
    const sections = [];
    for (const [key, items] of groupMap) {
        const groupEntry = items.find((i) => i.kind === "group");
        const title = groupEntry ? (pages.get(groupEntry.slug)?.title ?? key) : key;
        const types = items.filter((i) => i.kind !== "group" && i.kind !== "namespace");
        if (types.length === 0 && !groupEntry)
            continue;
        const links = [];
        if (groupEntry) {
            links.push(`- [Overview](${groupEntry.slug}.html)`);
        }
        for (const t of types) {
            links.push(`- [${t.title}](${t.slug}.html)`);
        }
        sections.push(`### ${title}\n\n${links.join("\n")}`);
    }
    return renderMarkdown(sections.join("\n\n"));
}
/**
 * Flat: alphabetical list categorized by kind.
 */
function buildFlatIndex(groupMap) {
    const byKind = new Map();
    for (const [, items] of groupMap) {
        for (const item of items) {
            if (item.kind === "group" || item.kind === "namespace")
                continue;
            const label = kindLabel(item.kind);
            if (!byKind.has(label))
                byKind.set(label, []);
            byKind.get(label).push(item);
        }
    }
    const sections = [];
    for (const [label, items] of byKind) {
        const sorted = items.sort((a, b) => a.title.localeCompare(b.title));
        const links = sorted.map((i) => `- [${i.title}](${i.slug}.html)`).join("\n");
        sections.push(`### ${label}\n\n${links}`);
    }
    return renderMarkdown(sections.join("\n\n"));
}
// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------
function lastSegment(name) {
    if (!name)
        return undefined;
    const parts = name.split("::");
    return parts[parts.length - 1] || name;
}
function escHtml(s) {
    return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}
function kindLabel(kind) {
    switch (kind) {
        case "class": return "Classes";
        case "struct": return "Structs";
        case "enum": return "Enums";
        case "union": return "Unions";
        case "typedef": return "Type Aliases";
        default: return kind.charAt(0).toUpperCase() + kind.slice(1) + "s";
    }
}
