// sourcey/dist/core/loader.js
import { readFile, access } from "node:fs/promises";
import { resolve, extname } from "node:path";
import yaml from "js-yaml";
/**
 * Load an OpenAPI/Swagger spec from a local file path or URL.
 * Auto-detects JSON vs YAML and Swagger 2.0 vs OpenAPI 3.x.
 */
export async function loadSpec(source) {
    const content = await fetchContent(source);
    const format = detectFormat(source, content);
    const raw = parseContent(content, format);
    const version = detectVersion(raw);
    const resolvedSource = isUrl(source) ? source : resolve(source);
    return { raw, format, version, source: resolvedSource };
}
/**
 * Fetch raw content from a file path or URL.
 */
async function fetchContent(source) {
    if (isUrl(source)) {
        const response = await fetch(source);
        if (!response.ok) {
            throw new Error(`Failed to fetch spec from ${source}: ${response.status} ${response.statusText}`);
        }
        return response.text();
    }
    const filePath = resolve(source);
    await access(filePath).catch(() => {
        throw new Error(`Spec file not found: ${filePath}`);
    });
    return readFile(filePath, "utf-8");
}
/**
 * Detect whether the content is JSON or YAML based on extension and content.
 */
function detectFormat(source, content) {
    const ext = extname(source).toLowerCase();
    if (ext === ".json")
        return "json";
    if (ext === ".yml" || ext === ".yaml")
        return "yaml";
    // Try to detect from content
    const trimmed = content.trimStart();
    if (trimmed.startsWith("{"))
        return "json";
    return "yaml";
}
/**
 * Parse raw string content into a JS object.
 */
function parseContent(content, format) {
    if (format === "json") {
        try {
            return JSON.parse(content);
        }
        catch (e) {
            throw new Error(`Failed to parse JSON spec: ${e.message}`);
        }
    }
    try {
        const parsed = yaml.load(content);
        if (typeof parsed !== "object" || parsed === null) {
            throw new Error("YAML spec must be an object");
        }
        return parsed;
    }
    catch (e) {
        if (e instanceof yaml.YAMLException) {
            throw new Error(`Failed to parse YAML spec: ${e.message}`);
        }
        throw e;
    }
}
/**
 * Detect the spec version from the parsed document.
 */
function detectVersion(raw) {
    if (typeof raw.swagger === "string" && raw.swagger.startsWith("2.")) {
        return "swagger-2.0";
    }
    if (typeof raw.openapi === "string") {
        if (raw.openapi.startsWith("3.1"))
            return "openapi-3.1";
        if (raw.openapi.startsWith("3."))
            return "openapi-3.0";
    }
    throw new Error('Unable to detect spec version. Expected "swagger": "2.0" or "openapi": "3.x.x"');
}
function isUrl(source) {
    return source.startsWith("http://") || source.startsWith("https://");
}
