// sourcey/dist/core/markdown-loader.js
import { readFile } from "node:fs/promises";
import { basename, extname, relative } from "node:path";
import { load as parseYaml } from "js-yaml";
import { htmlId } from "../utils/html-id.js";
import { renderIcon } from "../utils/icons.js";
import { renderCodeBlock, renderMarkdown, renderMarkdownInline, extractHeadings, } from "../utils/markdown.js";
// ---------------------------------------------------------------------------
// Frontmatter parsing
// ---------------------------------------------------------------------------
const FRONTMATTER_RE = /^---\s*\n([\s\S]*?)\n---\s*\n([\s\S]*)$/;
function parseFrontmatter(raw) {
    const match = raw.match(FRONTMATTER_RE);
    if (!match) {
        return { meta: {}, body: raw };
    }
    const meta = parseYaml(match[1]) ?? {};
    return { meta, body: match[2] };
}
const FENCED_BLOCK_TOKEN = "@@SOURCEY_FENCED_BLOCK_";
const INLINE_CODE_TOKEN = "@@SOURCEY_INLINE_CODE_";
/** Current page's protected fence blocks — used to restore placeholders in JSX children. */
let activeFenceBlocks = [];
/** Current page's protected inline code spans — used to restore placeholders in JSX children. */
let activeInlineSpans = [];
function protectFencedCodeBlocks(input) {
    const blocks = [];
    const output = [];
    const lines = input.split("\n");
    let fence = null;
    let buffer = [];
    const pushProtectedBlock = () => {
        const index = blocks.push(buffer.join("\n")) - 1;
        output.push(`${FENCED_BLOCK_TOKEN}${index}@@`);
        buffer = [];
    };
    for (const line of lines) {
        const stripped = line.replace(/^ {1,3}/, "");
        const fenceMatch = stripped.match(/^(`{3,}|~{3,})(.*)$/);
        if (!fence) {
            if (!fenceMatch) {
                output.push(line);
                continue;
            }
            fence = { char: fenceMatch[1][0], length: fenceMatch[1].length };
            buffer.push(line);
            continue;
        }
        buffer.push(line);
        if (fenceMatch &&
            fenceMatch[1][0] === fence.char &&
            fenceMatch[1].length >= fence.length &&
            !fenceMatch[2].trim()) {
            pushProtectedBlock();
            fence = null;
        }
    }
    if (buffer.length > 0) {
        output.push(...buffer);
    }
    return { text: output.join("\n"), blocks };
}
function restoreFencedCodeBlocks(input, blocks) {
    return blocks.reduce((text, block, index) => text.replaceAll(`${FENCED_BLOCK_TOKEN}${index}@@`, block), input);
}
function protectInlineCodeSpans(input) {
    const spans = [];
    let output = "";
    let i = 0;
    while (i < input.length) {
        if (input[i] !== "`") {
            output += input[i];
            i += 1;
            continue;
        }
        let tickEnd = i + 1;
        while (tickEnd < input.length && input[tickEnd] === "`")
            tickEnd += 1;
        const delimiter = input.slice(i, tickEnd);
        const closeIndex = input.indexOf(delimiter, tickEnd);
        if (closeIndex === -1) {
            output += input.slice(i);
            break;
        }
        const span = input.slice(i, closeIndex + delimiter.length);
        // If the content between delimiters spans multiple lines, it's a fenced
        // code block (or block-level backticks), not an inline code span — skip it.
        const inner = input.slice(tickEnd, closeIndex);
        if (inner.includes("\n")) {
            output += input[i];
            i += 1;
            continue;
        }
        const index = spans.push(span) - 1;
        output += `${INLINE_CODE_TOKEN}${index}@@`;
        i = closeIndex + delimiter.length;
    }
    return { text: output, spans };
}
function restoreInlineCodeSpans(input, spans) {
    return spans.reduce((text, span, index) => text.replaceAll(`${INLINE_CODE_TOKEN}${index}@@`, span), input);
}
function skipWhitespace(input, index) {
    let i = index;
    while (i < input.length && /\s/.test(input[i]))
        i += 1;
    return i;
}
function parseQuotedAttrValue(input, index) {
    const quote = input[index];
    if (quote !== `"` && quote !== `'`)
        return null;
    let value = "";
    let i = index + 1;
    while (i < input.length) {
        const ch = input[i];
        if (ch === "\\") {
            if (i + 1 < input.length) {
                value += input[i + 1];
                i += 2;
                continue;
            }
            i += 1;
            continue;
        }
        if (ch === quote) {
            return { value, nextIndex: i + 1 };
        }
        value += ch;
        i += 1;
    }
    return null;
}
function parseBracedAttrValue(input, index) {
    if (input[index] !== "{")
        return null;
    let depth = 1;
    let i = index + 1;
    while (i < input.length) {
        const ch = input[i];
        if (ch === `"` || ch === `'`) {
            const quoted = parseQuotedAttrValue(input, i);
            if (!quoted)
                return null;
            i = quoted.nextIndex;
            continue;
        }
        if (ch === "{") {
            depth += 1;
        }
        else if (ch === "}") {
            depth -= 1;
            if (depth === 0) {
                return {
                    value: input.slice(index + 1, i).trim(),
                    nextIndex: i + 1,
                };
            }
        }
        i += 1;
    }
    return null;
}
function parseKeyValueAttrs(raw, options) {
    const attrs = {};
    let i = 0;
    while (i < raw.length) {
        i = skipWhitespace(raw, i);
        if (i >= raw.length)
            break;
        const keyMatch = raw.slice(i).match(/^([A-Za-z_][A-Za-z0-9_-]*)/);
        if (!keyMatch)
            break;
        const key = keyMatch[1];
        i += key.length;
        i = skipWhitespace(raw, i);
        if (raw[i] !== "=")
            break;
        i += 1;
        i = skipWhitespace(raw, i);
        const parsedValue = raw[i] === "{" && options.allowBraces
            ? parseBracedAttrValue(raw, i)
            : parseQuotedAttrValue(raw, i);
        if (!parsedValue)
            break;
        attrs[key] = parsedValue.value;
        i = parsedValue.nextIndex;
    }
    return attrs;
}
const SUPPORTED_COMPONENT_TAGS = new Set([
    "Steps",
    "Step",
    "CardGroup",
    "Card",
    "AccordionGroup",
    "Accordion",
    "Expandable",
    "Tabs",
    "Tab",
    "CodeGroup",
    "Note",
    "Warning",
    "Tip",
    "Info",
    "Video",
    "Iframe",
]);
const SELF_CLOSING_COMPONENT_TAGS = new Set(["Video", "Iframe"]);
function escapeDirectiveAttr(value) {
    return value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}
function buildDirectiveAttrList(entries) {
    const attrs = entries
        .filter(([, value]) => value !== undefined && value !== "")
        .map(([key, value]) => `${key}="${escapeDirectiveAttr(value)}"`);
    return attrs.length > 0 ? `{${attrs.join(" ")}}` : "";
}
function indentBlock(text, prefix) {
    const lines = text.split("\n");
    const result = [];
    let fence = null;
    for (const line of lines) {
        if (fence) {
            // Inside a fenced code block — don't indent
            result.push(line);
            if (closesFence(line, fence))
                fence = null;
        }
        else {
            const nextFence = isFenceStart(line);
            if (nextFence) {
                // Fence opener — don't indent so markdown recognises it
                fence = nextFence;
                result.push(line);
            }
            else {
                result.push(`${prefix}${line}`);
            }
        }
    }
    return result.join("\n");
}
function isWhitespaceOnlyText(node) {
    return node.kind === "text" && node.value.trim() === "";
}
function findComponentTagEnd(input, start) {
    let quote = null;
    let braceDepth = 0;
    let i = start + 1;
    while (i < input.length) {
        const ch = input[i];
        if (quote) {
            if (ch === "\\") {
                i += 2;
                continue;
            }
            if (ch === quote)
                quote = null;
            i += 1;
            continue;
        }
        if (ch === `"` || ch === `'`) {
            quote = ch;
            i += 1;
            continue;
        }
        if (ch === "{") {
            braceDepth += 1;
            i += 1;
            continue;
        }
        if (ch === "}" && braceDepth > 0) {
            braceDepth -= 1;
            i += 1;
            continue;
        }
        if (ch === ">" && braceDepth === 0) {
            return i + 1;
        }
        i += 1;
    }
    return null;
}
function tryParseComponentTag(input, start) {
    if (input[start] !== "<")
        return null;
    const end = findComponentTagEnd(input, start);
    if (end === null)
        return null;
    const raw = input.slice(start, end);
    const inner = input.slice(start + 1, end - 1).trim();
    if (!inner)
        return null;
    if (inner.startsWith("/")) {
        const closeName = inner.slice(1).trim();
        if (!SUPPORTED_COMPONENT_TAGS.has(closeName))
            return null;
        return {
            name: closeName,
            kind: "close",
            attrs: {},
            raw,
            start,
            end,
        };
    }
    const selfClosing = inner.endsWith("/");
    const content = selfClosing ? inner.slice(0, -1).trimEnd() : inner;
    const nameMatch = content.match(/^([A-Za-z][A-Za-z0-9]*)/);
    if (!nameMatch || !SUPPORTED_COMPONENT_TAGS.has(nameMatch[1])) {
        return null;
    }
    const name = nameMatch[1];
    const attrSource = content.slice(name.length);
    return {
        name,
        kind: selfClosing || SELF_CLOSING_COMPONENT_TAGS.has(name) ? "self" : "open",
        attrs: parseKeyValueAttrs(attrSource, { allowBraces: true }),
        raw,
        start,
        end,
    };
}
function parseComponentNodes(input, startIndex = 0, untilTag) {
    const nodes = [];
    let cursor = startIndex;
    while (cursor < input.length) {
        const nextTag = input.indexOf("<", cursor);
        if (nextTag === -1) {
            if (cursor < input.length) {
                nodes.push({ kind: "text", value: input.slice(cursor) });
            }
            return { nodes, nextIndex: input.length };
        }
        if (nextTag > cursor) {
            nodes.push({ kind: "text", value: input.slice(cursor, nextTag) });
        }
        const tag = tryParseComponentTag(input, nextTag);
        if (!tag) {
            nodes.push({ kind: "text", value: "<" });
            cursor = nextTag + 1;
            continue;
        }
        if (tag.kind === "close") {
            if (untilTag && tag.name === untilTag) {
                return {
                    nodes,
                    nextIndex: tag.end,
                    matchedClose: tag,
                };
            }
            nodes.push({ kind: "text", value: tag.raw });
            cursor = tag.end;
            continue;
        }
        if (tag.kind === "self") {
            nodes.push({
                kind: "component",
                name: tag.name,
                attrs: tag.attrs,
                children: [],
                original: tag.raw,
            });
            cursor = tag.end;
            continue;
        }
        const childResult = parseComponentNodes(input, tag.end, tag.name);
        if (!childResult.matchedClose) {
            nodes.push({
                kind: "text",
                value: input.slice(nextTag, childResult.nextIndex),
            });
            cursor = childResult.nextIndex;
            continue;
        }
        nodes.push({
            kind: "component",
            name: tag.name,
            attrs: tag.attrs,
            children: childResult.nodes,
            original: input.slice(nextTag, childResult.nextIndex),
        });
        cursor = childResult.nextIndex;
    }
    return { nodes, nextIndex: cursor };
}
function renderParsedComponentNodes(nodes) {
    return nodes.map(renderParsedComponentNode).join("");
}
function collectChildComponents(children, expected) {
    const matches = [];
    for (const child of children) {
        if (isWhitespaceOnlyText(child))
            continue;
        if (child.kind !== "component" || child.name !== expected) {
            return null;
        }
        matches.push(child);
    }
    return matches;
}
/**
 * Render the raw text content of parsed component children through the full
 * markdown pipeline (component preprocessing → directive preprocessing →
 * markdown rendering).  Strips common leading whitespace first so that
 * indented JSX content (e.g. code fences inside <Step>) is handled correctly.
 */
function renderComponentChildrenToHtml(children) {
    // Save outer protection maps — renderDirectiveMarkdown → preprocessComponents will overwrite them
    const savedFenceBlocks = activeFenceBlocks;
    const savedInlineSpans = activeInlineSpans;
    let raw = renderParsedComponentNodes(children);
    raw = restoreFencedCodeBlocks(raw, savedFenceBlocks);
    raw = restoreInlineCodeSpans(raw, savedInlineSpans);
    const result = renderDirectiveMarkdown(dedent(raw));
    // Restore outer maps for sibling components
    activeFenceBlocks = savedFenceBlocks;
    activeInlineSpans = savedInlineSpans;
    return result;
}
/**
 * Strip common leading whitespace from a block of text.
 */
function dedent(text) {
    const lines = text.split("\n");
    const nonEmpty = lines.filter((l) => l.trim().length > 0);
    if (nonEmpty.length === 0)
        return text;
    const indent = Math.min(...nonEmpty.map((l) => l.match(/^(\s*)/)?.[1].length ?? 0));
    if (indent === 0)
        return text;
    return lines.map((l) => l.slice(indent)).join("\n");
}
/**
 * Render an accordion from a parsed JSX component directly to HTML.
 */
function renderComponentAccordion(node) {
    const title = renderMarkdownInline(node.attrs.title || "").trim();
    const body = renderComponentChildrenToHtml(node.children);
    return `<details class="accordion-item">
<summary class="accordion-trigger">${title}</summary>
<div class="accordion-content">
${body}
</div>
</details>`;
}
/**
 * Render a parsed JSX component directly to final HTML.
 * Returns null if the component structure is invalid (falls back to raw text).
 */
function renderParsedComponentElement(node) {
    if (node.name === "Steps") {
        const steps = collectChildComponents(node.children, "Step");
        if (!steps)
            return null;
        const items = steps
            .map((step, index) => {
            const title = renderMarkdownInline(step.attrs.title || "").trim();
            const body = renderComponentChildrenToHtml(step.children);
            return `<div role="listitem" class="step-item">
<div class="step-number">${index + 1}</div>
<div class="step-body">
<p class="step-title">${title}</p>
<div class="step-content">
${body}
</div>
</div>
</div>`;
        })
            .join("\n");
        return `\n\n<div role="list" class="steps not-prose">\n${items}\n</div>\n`;
    }
    if (node.name === "CardGroup") {
        const cards = collectChildComponents(node.children, "Card");
        if (!cards)
            return null;
        const cols = node.attrs.cols || "2";
        const cardHtml = cards
            .map((card) => {
            const tag = card.attrs.href ? "a" : "div";
            const href = card.attrs.href ? ` href="${escapeHtmlAttr(card.attrs.href)}"` : "";
            const iconHtml = renderIcon(card.attrs.icon || "");
            const title = renderMarkdownInline(card.attrs.title || "").trim();
            const body = renderComponentChildrenToHtml(card.children);
            return `<${tag}${href} class="card-item">
<div class="card-item-inner">
${iconHtml}
<h3 class="card-item-title">${title}</h3>
<div class="card-item-content">
${body}
</div>
</div>
</${tag}>`;
        })
            .join("\n");
        return `\n\n<div class="card-group not-prose" data-cols="${escapeHtmlAttr(cols)}">\n${cardHtml}\n</div>\n`;
    }
    if (node.name === "AccordionGroup") {
        const accordions = collectChildComponents(node.children, "Accordion");
        if (!accordions)
            return null;
        return `\n\n<div class="accordion-group not-prose">\n${accordions.map(renderComponentAccordion).join("\n")}\n</div>\n`;
    }
    if (node.name === "Accordion" || node.name === "Expandable") {
        return `\n\n${renderComponentAccordion(node)}\n`;
    }
    if (node.name === "Tabs") {
        const tabs = collectChildComponents(node.children, "Tab");
        if (!tabs)
            return null;
        const tabData = tabs.map((tab) => ({
            title: renderMarkdownInline(tab.attrs.title || "").trim(),
            body: renderComponentChildrenToHtml(tab.children),
        }));
        return `\n\n${buildTabbedHtml(tabData, nextId("tabs"))}\n`;
    }
    if (node.name === "CodeGroup") {
        let cgRaw = renderParsedComponentNodes(node.children);
        cgRaw = restoreFencedCodeBlocks(cgRaw, activeFenceBlocks);
        cgRaw = restoreInlineCodeSpans(cgRaw, activeInlineSpans);
        const codeBlocks = parseTitledCodeBlocks(dedent(cgRaw)).map((block) => ({
            title: renderMarkdownInline(block.title).trim(),
            body: renderCodeBlock(block.body, block.lang),
        }));
        if (codeBlocks.length === 0)
            return null;
        return `\n\n${buildTabbedHtml(codeBlocks, nextId("cg"), "directive-code-group")}\n`;
    }
    if (node.name === "Note" || node.name === "Warning" || node.name === "Tip" || node.name === "Info") {
        const type = node.name.toLowerCase();
        const label = renderMarkdownInline(node.attrs.title?.trim() || node.name.charAt(0).toUpperCase() + node.name.slice(1)).trim();
        const body = renderComponentChildrenToHtml(node.children);
        return `\n\n<div class="callout callout-${type} not-prose">
<div class="callout-title">${label}</div>
${body ? `<div class="callout-content">\n${body}\n</div>` : ""}
</div>\n`;
    }
    if (node.name === "Video") {
        const titleAttr = node.attrs.title ? `{title="${escapeDirectiveAttr(node.attrs.title)}"}` : "";
        return `::video[${node.attrs.src || ""}]${titleAttr}`;
    }
    if (node.name === "Iframe") {
        const attrString = buildDirectiveAttrList([
            ["title", node.attrs.title],
            ["height", node.attrs.height],
        ]);
        return `::iframe[${node.attrs.src || ""}]${attrString}`;
    }
    return null;
}
function renderParsedComponentNode(node) {
    if (node.kind === "text")
        return node.value;
    return renderParsedComponentElement(node) ?? node.original;
}
/**
 * Convert JSX-style components to directive syntax so they go through
 * a single rendering path in preprocessDirectives.
 */
function preprocessComponents(body) {
    const { text: fencedText, blocks } = protectFencedCodeBlocks(body);
    activeFenceBlocks = blocks;
    const { text, spans } = protectInlineCodeSpans(fencedText);
    activeInlineSpans = spans;
    const parsed = parseComponentNodes(text);
    const restoredInline = restoreInlineCodeSpans(renderParsedComponentNodes(parsed.nodes), spans);
    return restoreFencedCodeBlocks(restoredInline, blocks);
}
// ---------------------------------------------------------------------------
// Directive preprocessor: transforms :::directive blocks into HTML
// ---------------------------------------------------------------------------
/** Parse {key="value" key2="value2"} attribute strings. */
function parseAttrs(raw) {
    return parseKeyValueAttrs(raw, { allowBraces: false });
}
/** Deterministic tab-group ID; reset per page. */
let directiveCounter = 0;
function nextId(prefix) {
    return `${prefix}-${++directiveCounter}`;
}
/** Reset counter between builds (called from loadMarkdownPage). */
function resetDirectiveCounter() {
    directiveCounter = 0;
}
function escapeRegExp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function escapeHtmlAttr(value) {
    return value
        .replace(/&/g, "&amp;")
        .replace(/"/g, "&quot;")
        .replace(/</g, "&lt;");
}
function stripDirectiveIndent(line) {
    return line.replace(/^ {0,3}/, "");
}
function isFenceStart(line) {
    const match = stripDirectiveIndent(line).match(/^(`{3,}|~{3,})(.*)$/);
    if (!match)
        return null;
    return { char: match[1][0], length: match[1].length };
}
function closesFence(line, fence) {
    const match = stripDirectiveIndent(line).match(/^(`{3,}|~{3,})(.*)$/);
    return !!(match &&
        match[1][0] === fence.char &&
        match[1].length >= fence.length &&
        !match[2].trim());
}
/** Build tabbed UI HTML (shared by :::tabs and :::code-group). */
function buildTabbedHtml(tabs, id, extraClass) {
    const cls = extraClass
        ? `directive-tabs ${extraClass} not-prose`
        : "directive-tabs not-prose";
    const buttons = tabs
        .map((t, i) => `<button class="directive-tab${i === 0 ? " active" : ""}" data-tab-group="${id}" data-tab-index="${i}">${t.title}</button>`)
        .join("\n");
    const panels = tabs
        .map((t, i) => `<div class="directive-tab-panel${i === 0 ? " active" : ""}" data-tab-group="${id}" data-tab-index="${i}">

${t.body}

</div>`)
        .join("\n");
    return `<div class="${cls}">
<div class="directive-tab-bar">${buttons}</div>
${panels}
</div>`;
}
/**
 * Split content on `::child` markers, returning { attrs, body } for each.
 * Leading content before the first marker is discarded.
 */
function splitChildren(content, marker) {
    const children = [];
    const lines = content.split("\n");
    const startRe = new RegExp(`^::${escapeRegExp(marker)}(?:\\{([^}]*)\\})?\\s*$`);
    let i = 0;
    while (i < lines.length) {
        const start = stripDirectiveIndent(lines[i]).trimEnd().match(startRe);
        if (!start) {
            i += 1;
            continue;
        }
        let fence = null;
        let depth = 1;
        let j = i + 1;
        while (j < lines.length) {
            const line = lines[j];
            if (fence) {
                if (closesFence(line, fence))
                    fence = null;
                j += 1;
                continue;
            }
            const nextFence = isFenceStart(line);
            if (nextFence) {
                fence = nextFence;
                j += 1;
                continue;
            }
            const stripped = stripDirectiveIndent(line).trimEnd();
            if (startRe.test(stripped)) {
                depth += 1;
            }
            else if (/^::\s*$/.test(stripped)) {
                depth -= 1;
                if (depth === 0)
                    break;
            }
            j += 1;
        }
        if (depth !== 0) {
            i += 1;
            continue;
        }
        children.push({
            attrs: parseAttrs(start[1] ?? ""),
            body: lines.slice(i + 1, j).join("\n").trim(),
        });
        i = j + 1;
    }
    return children;
}
function parseTitledCodeBlocks(content) {
    const blocks = [];
    const lines = content.split("\n");
    let i = 0;
    while (i < lines.length) {
        const line = stripDirectiveIndent(lines[i]).trimEnd();
        const open = line.match(/^(`{3,}|~{3,})(\S+)?(?:\s+title="([^"]*)"|\s+(.+?))?\s*$/);
        if (!open) {
            i += 1;
            continue;
        }
        const fence = { char: open[1][0], length: open[1].length };
        let j = i + 1;
        while (j < lines.length && !closesFence(lines[j], fence))
            j += 1;
        if (j >= lines.length)
            break;
        blocks.push({
            title: open[3] ?? open[4] ?? "",
            lang: open[2] ?? "",
            body: lines.slice(i + 1, j).join("\n"),
        });
        i = j + 1;
    }
    return blocks;
}
const SUPPORTED_DIRECTIVES = new Set([
    "note",
    "warning",
    "tip",
    "info",
    "steps",
    "tabs",
    "code-group",
    "card-group",
    "accordion",
]);
function matchDirectiveStart(line) {
    const match = stripDirectiveIndent(line).trimEnd().match(/^:::(\w[\w-]*)(.*)$/);
    if (!match || !SUPPORTED_DIRECTIVES.has(match[1]))
        return null;
    return {
        type: match[1],
        meta: match[2].trim(),
    };
}
function isAnyDirectiveStart(line) {
    return /^:::\w[\w-]*(?:\{[^}]*\}|.*)?$/.test(stripDirectiveIndent(line).trimEnd());
}
function renderDirectiveMarkdown(content) {
    const trimmed = content.trim();
    if (!trimmed)
        return "";
    return renderMarkdown(preprocessDirectives(preprocessComponents(trimmed))).trim();
}
function collectDirectiveBlock(lines, startIndex) {
    let fence = null;
    let depth = 1;
    let i = startIndex + 1;
    while (i < lines.length) {
        const line = lines[i];
        if (fence) {
            if (closesFence(line, fence))
                fence = null;
            i += 1;
            continue;
        }
        const nextFence = isFenceStart(line);
        if (nextFence) {
            fence = nextFence;
            i += 1;
            continue;
        }
        const stripped = stripDirectiveIndent(line).trimEnd();
        if (isAnyDirectiveStart(stripped)) {
            depth += 1;
        }
        else if (/^:::\s*$/.test(stripped)) {
            depth -= 1;
            if (depth === 0) {
                return {
                    content: lines.slice(startIndex + 1, i).join("\n"),
                    nextLine: i + 1,
                };
            }
        }
        i += 1;
    }
    return null;
}
function renderDirectiveBlock(type, meta, content) {
    if (type === "note" || type === "warning" || type === "tip" || type === "info") {
        const label = renderMarkdownInline(meta.trim() || type.charAt(0).toUpperCase() + type.slice(1)).trim();
        const body = renderDirectiveMarkdown(content);
        return `\n\n<div class="callout callout-${type} not-prose">
<div class="callout-title">${label}</div>
${body ? `<div class="callout-content">\n${body}\n</div>` : ""}
</div>\n`;
    }
    if (type === "steps") {
        const lines = content.trim().split("\n");
        const steps = [];
        let fence = null;
        for (const line of lines) {
            if (fence) {
                if (closesFence(line, fence))
                    fence = null;
            }
            else {
                const nextFence = isFenceStart(line);
                if (nextFence) {
                    fence = nextFence;
                }
            }
            const stripped = line.replace(/^ {1,3}/, "");
            const stepMatch = !fence && line.match(/^\d+\.\s+(.+)/);
            if (stepMatch) {
                steps.push({ title: stepMatch[1], body: [] });
            }
            else if (steps.length > 0) {
                steps[steps.length - 1].body.push(stripped);
            }
        }
        if (steps.length === 0)
            return renderDirectiveMarkdown(content);
        const items = steps
            .map((step, index) => {
            const title = renderMarkdownInline(step.title).trim();
            const body = renderDirectiveMarkdown(step.body.join("\n").trim());
            return `<div role="listitem" class="step-item">
<div class="step-number">${index + 1}</div>
<div class="step-body">
<p class="step-title">${title}</p>
<div class="step-content">
${body}
</div>
</div>
</div>`;
        })
            .join("\n");
        return `\n\n<div role="list" class="steps not-prose">\n${items}\n</div>\n`;
    }
    if (type === "tabs") {
        const tabs = splitChildren(content, "tab").map((child) => ({
            title: renderMarkdownInline(child.attrs.title || "").trim(),
            body: renderDirectiveMarkdown(child.body),
        }));
        if (tabs.length === 0)
            return renderDirectiveMarkdown(content);
        return `\n\n${buildTabbedHtml(tabs, nextId("tabs"))}\n`;
    }
    if (type === "code-group") {
        const codeBlocks = parseTitledCodeBlocks(content).map((block) => ({
            title: renderMarkdownInline(block.title).trim(),
            body: renderCodeBlock(block.body, block.lang),
        }));
        if (codeBlocks.length === 0)
            return renderDirectiveMarkdown(content);
        return `\n\n${buildTabbedHtml(codeBlocks, nextId("cg"), "directive-code-group")}\n`;
    }
    if (type === "card-group") {
        const cols = parseAttrs(meta.match(/^\{([^}]*)\}$/)?.[1] ?? "").cols || "2";
        const cards = splitChildren(content, "card").map((child) => {
            const tag = child.attrs.href ? "a" : "div";
            const href = child.attrs.href ? ` href="${escapeHtmlAttr(child.attrs.href)}"` : "";
            const iconHtml = renderIcon(child.attrs.icon || "");
            const title = renderMarkdownInline(child.attrs.title || "").trim();
            const body = renderDirectiveMarkdown(child.body);
            return `<${tag}${href} class="card-item">
<div class="card-item-inner">
${iconHtml}
<h3 class="card-item-title">${title}</h3>
<div class="card-item-content">
${body}
</div>
</div>
</${tag}>`;
        });
        if (cards.length === 0)
            return renderDirectiveMarkdown(content);
        return `\n\n<div class="card-group not-prose" data-cols="${escapeHtmlAttr(cols)}">\n${cards.join("\n")}\n</div>\n`;
    }
    const title = renderMarkdownInline(parseAttrs(meta.match(/^\{([^}]*)\}$/)?.[1] ?? "").title || "").trim();
    const body = renderDirectiveMarkdown(content);
    return `\n\n<details class="accordion-item">
<summary class="accordion-trigger">${title}</summary>
<div class="accordion-content">
${body}
</div>
</details>\n`;
}
/**
 * Process :::directive blocks. Handles callouts, steps, tabs,
 * code-group, card-group, and accordion.
 *
 * Matching strategy: process one block type at a time. Each regex
 * matches a specific `:::type` opener so closing `:::` markers
 * from other directive types cannot collide.
 */
function preprocessDirectives(body) {
    const lines = body.split("\n");
    const out = [];
    let i = 0;
    let fence = null;
    while (i < lines.length) {
        if (fence) {
            out.push(lines[i]);
            if (closesFence(lines[i], fence))
                fence = null;
            i += 1;
            continue;
        }
        const nextFence = isFenceStart(lines[i]);
        if (nextFence) {
            fence = nextFence;
            out.push(lines[i]);
            i += 1;
            continue;
        }
        const start = matchDirectiveStart(lines[i]);
        if (!start) {
            out.push(lines[i]);
            i += 1;
            continue;
        }
        if (start.type === "accordion") {
            const accordions = [];
            let nextLine = i;
            while (true) {
                const accordionStart = matchDirectiveStart(lines[nextLine]);
                if (!accordionStart || accordionStart.type !== "accordion")
                    break;
                const accordionBlock = collectDirectiveBlock(lines, nextLine);
                if (!accordionBlock)
                    break;
                accordions.push(renderDirectiveBlock("accordion", accordionStart.meta, accordionBlock.content));
                nextLine = accordionBlock.nextLine;
                let scan = nextLine;
                while (scan < lines.length && /^\s*$/.test(lines[scan]))
                    scan += 1;
                const nextAccordion = scan < lines.length ? matchDirectiveStart(lines[scan]) : null;
                if (!nextAccordion || nextAccordion.type !== "accordion")
                    break;
                nextLine = scan;
            }
            if (accordions.length > 1) {
                out.push(`\n\n<div class="accordion-group not-prose">\n${accordions.join("\n")}\n</div>\n`);
                i = nextLine;
                continue;
            }
        }
        const block = collectDirectiveBlock(lines, i);
        if (!block) {
            out.push(lines[i]);
            i += 1;
            continue;
        }
        out.push(renderDirectiveBlock(start.type, start.meta, block.content));
        i = block.nextLine;
    }
    return out.join("\n");
}
// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------
/**
 * Load a single markdown file and return a MarkdownPage.
 */
export async function loadMarkdownPage(filePath, slug) {
    const raw = await readFile(filePath, "utf-8");
    const { meta, body } = parseFrontmatter(raw);
    resetDirectiveCounter();
    const componentPreprocessed = preprocessComponents(body);
    const preprocessed = preprocessDirectives(componentPreprocessed);
    const headings = extractHeadings(componentPreprocessed);
    const html = renderMarkdown(preprocessed);
    const title = meta.title ?? extractFirstHeading(componentPreprocessed) ?? slug;
    const description = meta.description ?? "";
    // Strip leading h1 from rendered HTML when it duplicates the page title
    const cleanHtml = html.replace(/^\s*<h1[^>]*>(.*?)<\/h1>\s*/i, (_m, inner) => inner.replace(/<[^>]+>/g, "").trim() === title ? "" : _m);
    const sourcePath = relative(process.cwd(), filePath);
    return {
        title,
        description,
        slug,
        html: cleanHtml,
        headings,
        sourcePath,
        editPath: sourcePath,
    };
}
/**
 * Extract the first # heading from markdown as a fallback title.
 */
function extractFirstHeading(md) {
    const match = md.match(/^#\s+(.+)$/m);
    return match ? match[1].trim() : undefined;
}
/**
 * Derive a URL slug from a page path.
 * Preserves directory structure to avoid collisions:
 * "getting-started.md" → "getting-started"
 * "run/index"          → "run/index"
 * "run/install.md"     → "run/install"
 */
export function slugFromPath(filePath) {
    const ext = extname(filePath);
    const stripped = ext ? filePath.slice(0, -ext.length) : filePath;
    return stripped.split("/").map((s) => htmlId(s)).join("/");
}
