// sourcey/dist/core/mcp-normalizer.js
/**
 * Convert an McpSpec (from mcp-parser) into a NormalizedSpec for rendering.
 *
 * This is the bridge between the MCP snapshot format and sourcey's
 * format-agnostic rendering pipeline. Everything downstream (components,
 * search, navigation, themes) works unchanged.
 */
import { generateExample } from "../utils/example-generator.js";
// ── Public API ────────────────────────────────────────────────────────
export function normalizeMcpSpec(spec) {
    const connectionInfo = {
        transport: spec.transport ? {
            type: spec.transport.type,
            command: spec.transport.command,
            args: spec.transport.args,
            url: spec.transport.url,
        } : undefined,
        serverName: spec.server.name,
        mcpVersion: spec.mcpVersion,
        capabilities: spec.capabilities,
    };
    const toolOps = (spec.tools ?? []).map((t) => toolToOperation(t, connectionInfo));
    const resourceOps = (spec.resources ?? []).map((r) => resourceToOperation(r, connectionInfo));
    const templateOps = (spec.resourceTemplates ?? []).map((t) => resourceTemplateToOperation(t, connectionInfo));
    const promptOps = (spec.prompts ?? []).map((p) => promptToOperation(p, connectionInfo));
    const allOps = [...toolOps, ...resourceOps, ...templateOps, ...promptOps];
    // Build tags — only for non-empty groups
    const tags = [];
    if (toolOps.length)
        tags.push({ name: "Tools", operations: toolOps });
    if (resourceOps.length || templateOps.length)
        tags.push({ name: "Resources", operations: [...resourceOps, ...templateOps] });
    if (promptOps.length)
        tags.push({ name: "Prompts", operations: promptOps });
    // Shared schema definitions from $defs
    const schemas = {};
    if (spec.$defs) {
        for (const [name, schema] of Object.entries(spec.$defs)) {
            schemas[name] = convertSchema(schema, name);
        }
    }
    return {
        info: {
            title: spec.server.name,
            version: spec.server.version,
            description: spec.description,
        },
        servers: spec.transport?.url ? [{ url: spec.transport.url }] : [],
        tags,
        operations: allOps,
        schemas,
        securitySchemes: {},
        webhooks: [],
    };
}
// ── Tool → Operation ──────────────────────────────────────────────────
function toolToOperation(tool, connection) {
    const group = tool["x-sourcey-group"];
    const inputSchema = convertSchema(tool.inputSchema);
    const outputSchema = tool.outputSchema ? convertSchema(tool.outputSchema) : undefined;
    // Flat parameters from top-level inputSchema properties
    const parameters = flattenInputSchema(tool.inputSchema);
    // Only include requestBody for nested/complex schemas
    const requestBody = hasNestedProperties(tool.inputSchema) ? {
        required: true,
        content: { "application/json": { schema: inputSchema } },
    } : undefined;
    return {
        operationId: tool.name,
        summary: tool.title ?? tool.name,
        description: tool.description,
        method: "tool",
        path: tool.name,
        tags: [group ?? "Tools"],
        parameters,
        requestBody,
        responses: [],
        security: [],
        deprecated: false,
        codeSamples: generateToolSamples(tool, inputSchema),
        mcpExtras: {
            type: "tool",
            annotations: tool.annotations ? {
                readOnlyHint: tool.annotations.readOnlyHint,
                destructiveHint: tool.annotations.destructiveHint,
                idempotentHint: tool.annotations.idempotentHint,
                openWorldHint: tool.annotations.openWorldHint,
            } : undefined,
            outputSchema,
            connection,
        },
    };
}
// ── Resource → Operation ──────────────────────────────────────────────
function resourceToOperation(resource, connection) {
    const group = resource["x-sourcey-group"];
    return {
        operationId: slugify(resource.name),
        summary: resource.name,
        description: resource.description,
        method: "resource",
        path: resource.uri,
        tags: [group ?? "Resources"],
        parameters: [],
        responses: [],
        security: [],
        deprecated: false,
        codeSamples: generateResourceSamples(resource),
        mcpExtras: {
            type: "resource",
            connection,
        },
    };
}
// ── Resource Template → Operation ─────────────────────────────────────
function resourceTemplateToOperation(template, connection) {
    const group = template["x-sourcey-group"];
    return {
        operationId: slugify(template.name),
        summary: template.name,
        description: template.description,
        method: "resource",
        path: template.uriTemplate,
        tags: [group ?? "Resources"],
        parameters: extractUriTemplateParams(template.uriTemplate),
        responses: [],
        security: [],
        deprecated: false,
        codeSamples: generateResourceSamples({ uri: template.uriTemplate, name: template.name }),
        mcpExtras: {
            type: "resource",
            connection,
        },
    };
}
// ── Prompt → Operation ────────────────────────────────────────────────
function promptToOperation(prompt, connection) {
    const group = prompt["x-sourcey-group"];
    const parameters = (prompt.arguments ?? []).map(arg => ({
        name: arg.name,
        in: "argument",
        description: arg.description,
        required: arg.required ?? false,
        deprecated: false,
        schema: { type: "string" },
    }));
    return {
        operationId: prompt.name,
        summary: prompt.name,
        description: prompt.description,
        method: "prompt",
        path: prompt.name,
        tags: [group ?? "Prompts"],
        parameters,
        responses: [],
        security: [],
        deprecated: false,
        codeSamples: generatePromptSamples(prompt),
        mcpExtras: {
            type: "prompt",
            connection,
        },
    };
}
// ── Schema conversion ─────────────────────────────────────────────────
function convertSchema(schema, name) {
    const result = {};
    if (name)
        result.name = name;
    if (schema.type)
        result.type = schema.type;
    if (schema.format)
        result.format = schema.format;
    if (schema.title)
        result.title = schema.title;
    if (schema.description)
        result.description = schema.description;
    if (schema.default !== undefined)
        result.default = schema.default;
    if (schema.enum)
        result.enum = schema.enum;
    if (schema.const !== undefined)
        result.const = schema.const;
    // Object
    if (schema.properties) {
        result.properties = {};
        for (const [key, prop] of Object.entries(schema.properties)) {
            result.properties[key] = convertSchema(prop, key);
        }
    }
    if (schema.additionalProperties !== undefined) {
        result.additionalProperties = typeof schema.additionalProperties === "boolean"
            ? schema.additionalProperties
            : convertSchema(schema.additionalProperties);
    }
    if (schema.required)
        result.required = schema.required;
    // Array — handle JsonSchema.items being JsonSchema | JsonSchema[]
    if (schema.items) {
        if (Array.isArray(schema.items)) {
            result.items = schema.items.length === 1
                ? convertSchema(schema.items[0])
                : { oneOf: schema.items.map(s => convertSchema(s)) };
        }
        else {
            result.items = convertSchema(schema.items);
        }
    }
    if (schema.minItems !== undefined)
        result.minItems = schema.minItems;
    if (schema.maxItems !== undefined)
        result.maxItems = schema.maxItems;
    if (schema.uniqueItems !== undefined)
        result.uniqueItems = schema.uniqueItems;
    // Numeric
    if (schema.minimum !== undefined)
        result.minimum = schema.minimum;
    if (schema.maximum !== undefined)
        result.maximum = schema.maximum;
    if (schema.exclusiveMinimum !== undefined)
        result.exclusiveMinimum = schema.exclusiveMinimum;
    if (schema.exclusiveMaximum !== undefined)
        result.exclusiveMaximum = schema.exclusiveMaximum;
    if (schema.multipleOf !== undefined)
        result.multipleOf = schema.multipleOf;
    // String
    if (schema.minLength !== undefined)
        result.minLength = schema.minLength;
    if (schema.maxLength !== undefined)
        result.maxLength = schema.maxLength;
    if (schema.pattern)
        result.pattern = schema.pattern;
    // Composition
    if (schema.allOf)
        result.allOf = schema.allOf.map((s) => convertSchema(s));
    if (schema.anyOf)
        result.anyOf = schema.anyOf.map((s) => convertSchema(s));
    if (schema.oneOf)
        result.oneOf = schema.oneOf.map((s) => convertSchema(s));
    if (schema.not)
        result.not = convertSchema(schema.not);
    // Examples
    if (schema.examples)
        result.examples = schema.examples;
    return result;
}
// ── Helpers ───────────────────────────────────────────────────────────
function flattenInputSchema(schema) {
    if (!schema.properties)
        return [];
    const required = new Set(schema.required ?? []);
    return Object.entries(schema.properties).map(([name, prop]) => ({
        name,
        in: "argument",
        description: prop.description,
        required: required.has(name),
        deprecated: false,
        schema: convertSchema(prop),
        example: prop.default,
    }));
}
function hasNestedProperties(schema) {
    if (!schema.properties)
        return false;
    return Object.values(schema.properties).some((prop) => {
        if (prop.type === "object" && prop.properties)
            return true;
        if (prop.type === "array" && prop.items && !Array.isArray(prop.items) && prop.items.type === "object")
            return true;
        if (prop.allOf || prop.anyOf || prop.oneOf)
            return true;
        return false;
    });
}
function extractUriTemplateParams(uriTemplate) {
    const matches = uriTemplate.match(/\{([^}]+)\}/g);
    if (!matches)
        return [];
    return matches.map(m => ({
        name: m.slice(1, -1),
        in: "path",
        required: true,
        deprecated: false,
        schema: { type: "string" },
    }));
}
function slugify(name) {
    return name.replace(/[^a-z0-9]+/gi, "-").toLowerCase();
}
// ── Code sample generation ────────────────────────────────────────────
function generateToolSamples(tool, inputSchema) {
    const example = inputSchema.properties
        ? generateExample(inputSchema)
        : {};
    const exampleJson = JSON.stringify(example, null, 2);
    const indented = exampleJson.split("\n").map((line, i) => i === 0 ? line : "      " + line).join("\n");
    return [
        {
            lang: "json",
            label: "JSON-RPC",
            source: `{
  "jsonrpc": "2.0",
  "method": "tools/call",
  "params": {
    "name": "${tool.name}",
    "arguments": ${indented}
  }
}`,
        },
        {
            lang: "typescript",
            label: "TypeScript",
            source: `const result = await client.callTool("${tool.name}", ${exampleJson});`,
        },
        {
            lang: "python",
            label: "Python",
            source: `result = await session.call_tool("${tool.name}", arguments=${exampleJson})`,
        },
    ];
}
function generateResourceSamples(resource) {
    return [
        {
            lang: "json",
            label: "JSON-RPC",
            source: `{
  "jsonrpc": "2.0",
  "method": "resources/read",
  "params": {
    "uri": "${resource.uri}"
  }
}`,
        },
        {
            lang: "typescript",
            label: "TypeScript",
            source: `const result = await client.readResource("${resource.uri}");`,
        },
        {
            lang: "python",
            label: "Python",
            source: `result = await session.read_resource("${resource.uri}")`,
        },
    ];
}
function generatePromptSamples(prompt) {
    const args = (prompt.arguments ?? []);
    const exampleArgs = Object.fromEntries(args.map((a) => [a.name, `<${a.name}>`]));
    const argsJson = JSON.stringify(exampleArgs, null, 2);
    const indented = argsJson.split("\n").map((line, i) => i === 0 ? line : "      " + line).join("\n");
    const tsArgs = args.map((a) => `  ${a.name}: "<${a.name}>",`).join("\n");
    const pyArgs = args.map((a) => `    "${a.name}": "<${a.name}>",`).join("\n");
    return [
        {
            lang: "json",
            label: "JSON-RPC",
            source: `{
  "jsonrpc": "2.0",
  "method": "prompts/get",
  "params": {
    "name": "${prompt.name}",
    "arguments": ${indented}
  }
}`,
        },
        {
            lang: "typescript",
            label: "TypeScript",
            source: args.length
                ? `const result = await client.getPrompt("${prompt.name}", {\n${tsArgs}\n});`
                : `const result = await client.getPrompt("${prompt.name}");`,
        },
        {
            lang: "python",
            label: "Python",
            source: args.length
                ? `result = await session.get_prompt("${prompt.name}", arguments={\n${pyArgs}\n})`
                : `result = await session.get_prompt("${prompt.name}")`,
        },
    ];
}
