// sourcey/dist/core/normalizer.js
const HTTP_METHODS = [
    "get", "post", "put", "delete", "patch", "options", "head", "trace",
];
/**
 * Transform a fully dereferenced OpenAPI 3.x document into a NormalizedSpec.
 * This is the single point where spec-format-specific logic lives.
 */
export function normalizeSpec(parsed) {
    const doc = parsed.document;
    const schemas = normalizeSchemas(doc.components?.schemas);
    const securitySchemes = normalizeSecuritySchemes(doc.components?.securitySchemes);
    const servers = normalizeServers(doc.servers);
    const operations = normalizeOperations(doc.paths, doc.security);
    const webhooks = normalizeWebhooks(doc.webhooks);
    const tags = normalizeTags(doc.tags, operations);
    return {
        info: normalizeInfo(doc),
        servers,
        tags,
        operations,
        schemas,
        securitySchemes,
        webhooks,
        externalDocs: normalizeExternalDocs(doc.externalDocs),
    };
}
// ── Info ───────────────────────────────────────────────────────────
function normalizeInfo(doc) {
    const info = doc.info ?? {};
    return {
        title: str(info.title, "Untitled API"),
        version: str(info.version, "0.0.0"),
        description: optStr(info.description),
        termsOfService: optStr(info.termsOfService),
        contact: info.contact
            ? {
                name: optStr(info.contact.name),
                url: optStr(info.contact.url),
                email: optStr(info.contact.email),
            }
            : undefined,
        license: info.license
            ? {
                name: str(info.license.name, ""),
                url: optStr(info.license.url),
                identifier: optStr(info.license.identifier),
            }
            : undefined,
        logo: optStr(info["x-logo"]),
    };
}
// ── Servers ────────────────────────────────────────────────────────
function normalizeServers(servers) {
    if (!servers?.length)
        return [{ url: "/" }];
    return servers.map((s) => ({
        url: str(s.url, "/"),
        description: optStr(s.description),
        variables: s.variables
            ? normalizeServerVariables(s.variables)
            : undefined,
    }));
}
function normalizeServerVariables(vars) {
    const result = {};
    for (const [key, v] of Object.entries(vars)) {
        result[key] = {
            default: str(v.default, ""),
            description: optStr(v.description),
            enum: Array.isArray(v.enum) ? v.enum.map(String) : undefined,
        };
    }
    return result;
}
// ── Tags ───────────────────────────────────────────────────────────
function normalizeTags(rawTags, operations) {
    // Build a map of defined tags
    const tagMap = new Map();
    if (rawTags) {
        for (const t of rawTags) {
            const name = str(t.name, "");
            if (name) {
                tagMap.set(name, {
                    name,
                    description: optStr(t.description),
                    externalDocs: normalizeExternalDocs(t.externalDocs),
                    operations: [],
                    hidden: t["x-sourcey-hide"] === true,
                });
            }
        }
    }
    // Assign operations to tags, creating implicit tags as needed
    for (const op of operations) {
        if (op.hidden)
            continue;
        const opTags = op.tags.length > 0 ? op.tags : ["default"];
        for (const tagName of opTags) {
            if (!tagMap.has(tagName)) {
                // Implicit tag creation (legacy behavior from preprocessor.js)
                tagMap.set(tagName, {
                    name: tagName,
                    operations: [],
                });
            }
            tagMap.get(tagName).operations.push(op);
        }
    }
    // Filter out tags with no operations (unless they have a description)
    return Array.from(tagMap.values()).filter((t) => t.operations.length > 0 || t.description);
}
// ── Operations ─────────────────────────────────────────────────────
function normalizeOperations(paths, globalSecurity) {
    if (!paths)
        return [];
    const operations = [];
    for (const [path, pathItem] of Object.entries(paths)) {
        const pathParams = normalizeParameters(pathItem.parameters);
        for (const method of HTTP_METHODS) {
            const opRaw = pathItem[method];
            if (!opRaw)
                continue;
            const opParams = normalizeParameters(opRaw.parameters);
            // Merge path-level + operation-level params, deduping by name+in
            const mergedParams = deduplicateParameters([...pathParams, ...opParams]);
            const op = {
                operationId: optStr(opRaw.operationId),
                summary: optStr(opRaw.summary),
                description: optStr(opRaw.description),
                method,
                path,
                tags: Array.isArray(opRaw.tags) ? opRaw.tags.map(String) : [],
                parameters: mergedParams,
                requestBody: normalizeRequestBody(opRaw.requestBody),
                responses: normalizeResponses(opRaw.responses),
                security: normalizeSecurityRequirements(opRaw.security, globalSecurity),
                deprecated: opRaw.deprecated === true,
                servers: opRaw.servers
                    ? normalizeServers(opRaw.servers)
                    : undefined,
                externalDocs: normalizeExternalDocs(opRaw.externalDocs),
                callbacks: normalizeCallbacks(opRaw.callbacks),
                hidden: opRaw["x-sourcey-hide"] === true,
                codeSamples: normalizeCodeSamples(opRaw["x-sourcey-code-samples"], opRaw["x-code-samples"]),
            };
            operations.push(op);
        }
    }
    return operations;
}
/**
 * Deduplicate parameters by (name, in) — operation-level params
 * override path-level params with the same identity.
 */
function deduplicateParameters(params) {
    const seen = new Map();
    // Process in order: path-level first, then operation-level.
    // Later entries (operation-level) override earlier ones (path-level).
    for (const p of params) {
        seen.set(`${p.in}:${p.name}`, p);
    }
    return Array.from(seen.values());
}
function normalizeParameters(params) {
    if (!params)
        return [];
    return params.map((p) => ({
        name: str(p.name, ""),
        in: str(p.in, "query"),
        description: optStr(p.description),
        required: p.required === true,
        deprecated: p.deprecated === true,
        schema: p.schema
            ? normalizeSchema(p.schema)
            : undefined,
        example: p.example,
        examples: p.examples
            ? normalizeExamples(p.examples)
            : undefined,
    }));
}
function normalizeRequestBody(body) {
    if (!body)
        return undefined;
    return {
        description: optStr(body.description),
        required: body.required === true,
        content: normalizeContent(body.content),
    };
}
function normalizeContent(content) {
    if (!content)
        return {};
    const result = {};
    for (const [mediaType, value] of Object.entries(content)) {
        result[mediaType] = {
            schema: value.schema
                ? normalizeSchema(value.schema)
                : undefined,
            example: value.example,
            examples: value.examples
                ? normalizeExamples(value.examples)
                : undefined,
        };
    }
    return result;
}
function normalizeResponses(responses) {
    if (!responses)
        return [];
    return Object.entries(responses).map(([statusCode, r]) => ({
        statusCode,
        description: str(r.description, ""),
        content: r.content
            ? normalizeContent(r.content)
            : undefined,
        headers: r.headers
            ? normalizeResponseHeaders(r.headers)
            : undefined,
        links: r.links
            ? normalizeLinks(r.links)
            : undefined,
    }));
}
function normalizeResponseHeaders(headers) {
    const result = {};
    for (const [name, h] of Object.entries(headers)) {
        result[name] = {
            name,
            in: "header",
            description: optStr(h.description),
            required: h.required === true,
            deprecated: h.deprecated === true,
            schema: h.schema
                ? normalizeSchema(h.schema)
                : undefined,
        };
    }
    return result;
}
function normalizeLinks(links) {
    const result = {};
    for (const [name, l] of Object.entries(links)) {
        result[name] = {
            operationId: optStr(l.operationId),
            operationRef: optStr(l.operationRef),
            description: optStr(l.description),
            parameters: l.parameters,
            requestBody: l.requestBody,
            server: l.server
                ? normalizeServers([l.server])[0]
                : undefined,
        };
    }
    return result;
}
function normalizeCallbacks(callbacks) {
    if (!callbacks)
        return undefined;
    const result = {};
    for (const [expression, pathItem] of Object.entries(callbacks)) {
        result[expression] = {
            expression,
            operations: normalizeOperations({ [expression]: pathItem }),
        };
    }
    return result;
}
function normalizeSecurityRequirements(opSecurity, globalSecurity) {
    // Operation security overrides global
    return (opSecurity ?? globalSecurity ?? []);
}
function normalizeCodeSamples(sourceySamples, genericSamples) {
    const samples = sourceySamples ?? genericSamples;
    if (!samples?.length)
        return undefined;
    return samples.map((s) => ({
        lang: str(s.lang, "text"),
        label: optStr(s.label),
        source: str(s.source, ""),
    }));
}
// ── Schemas ────────────────────────────────────────────────────────
function normalizeSchemas(schemas) {
    if (!schemas)
        return {};
    const result = {};
    for (const [name, s] of Object.entries(schemas)) {
        result[name] = normalizeSchema(s, name);
    }
    return result;
}
function normalizeSchema(schema, name) {
    // Handle nullable: OpenAPI 3.0 uses x-nullable or nullable,
    // OpenAPI 3.1 uses type: ["string", "null"]
    let type = schema.type;
    let nullable = schema.nullable === true || schema["x-nullable"] === true;
    if (Array.isArray(type)) {
        const nonNullTypes = type.filter((t) => t !== "null");
        if (nonNullTypes.length < type.length) {
            nullable = true;
        }
        type = nonNullTypes.length === 1 ? nonNullTypes[0] : nonNullTypes;
    }
    const result = {
        name,
        type,
        format: optStr(schema.format),
        title: optStr(schema.title),
        description: optStr(schema.description),
        default: schema.default,
        enum: Array.isArray(schema.enum) ? schema.enum : undefined,
        const: schema.const,
        nullable,
        readOnly: schema.readOnly === true,
        writeOnly: schema.writeOnly === true,
        deprecated: schema.deprecated === true,
        example: schema.example,
        examples: Array.isArray(schema.examples) ? schema.examples : undefined,
        refName: optStr(schema["x-ref-name"]),
    };
    // Object properties
    if (schema.properties) {
        result.properties = {};
        for (const [key, val] of Object.entries(schema.properties)) {
            result.properties[key] = normalizeSchema(val);
        }
    }
    if (schema.additionalProperties !== undefined) {
        result.additionalProperties =
            typeof schema.additionalProperties === "boolean"
                ? schema.additionalProperties
                : normalizeSchema(schema.additionalProperties);
    }
    if (Array.isArray(schema.required)) {
        result.required = schema.required.map(String);
    }
    // Array items
    if (schema.items) {
        result.items = normalizeSchema(schema.items);
    }
    if (schema.minItems !== undefined)
        result.minItems = Number(schema.minItems);
    if (schema.maxItems !== undefined)
        result.maxItems = Number(schema.maxItems);
    if (schema.uniqueItems === true)
        result.uniqueItems = true;
    // Numeric constraints
    if (schema.minimum !== undefined)
        result.minimum = Number(schema.minimum);
    if (schema.maximum !== undefined)
        result.maximum = Number(schema.maximum);
    if (schema.exclusiveMinimum !== undefined) {
        result.exclusiveMinimum = schema.exclusiveMinimum;
    }
    if (schema.exclusiveMaximum !== undefined) {
        result.exclusiveMaximum = schema.exclusiveMaximum;
    }
    if (schema.multipleOf !== undefined)
        result.multipleOf = Number(schema.multipleOf);
    // String constraints
    if (schema.minLength !== undefined)
        result.minLength = Number(schema.minLength);
    if (schema.maxLength !== undefined)
        result.maxLength = Number(schema.maxLength);
    if (schema.pattern !== undefined)
        result.pattern = String(schema.pattern);
    // Composition
    if (Array.isArray(schema.allOf)) {
        result.allOf = schema.allOf.map((s) => normalizeSchema(s));
    }
    if (Array.isArray(schema.anyOf)) {
        result.anyOf = schema.anyOf.map((s) => normalizeSchema(s));
    }
    if (Array.isArray(schema.oneOf)) {
        result.oneOf = schema.oneOf.map((s) => normalizeSchema(s));
    }
    if (schema.not) {
        result.not = normalizeSchema(schema.not);
    }
    if (schema.discriminator) {
        const d = schema.discriminator;
        result.discriminator = {
            propertyName: str(d.propertyName, ""),
            mapping: d.mapping,
        };
    }
    // External docs
    if (schema.externalDocs) {
        result.externalDocs = normalizeExternalDocs(schema.externalDocs);
    }
    return result;
}
// ── Security Schemes ───────────────────────────────────────────────
function normalizeSecuritySchemes(schemes) {
    if (!schemes)
        return {};
    const result = {};
    for (const [name, s] of Object.entries(schemes)) {
        const raw = s;
        result[name] = {
            type: str(raw.type, "apiKey"),
            name: optStr(raw.name),
            description: optStr(raw.description),
            in: optStr(raw.in),
            scheme: optStr(raw.scheme),
            bearerFormat: optStr(raw.bearerFormat),
            flows: raw.flows
                ? normalizeOAuthFlows(raw.flows)
                : undefined,
            openIdConnectUrl: optStr(raw.openIdConnectUrl),
        };
    }
    return result;
}
function normalizeOAuthFlows(flows) {
    const result = {};
    for (const flowType of [
        "implicit",
        "password",
        "clientCredentials",
        "authorizationCode",
    ]) {
        if (flows[flowType]) {
            result[flowType] = normalizeOAuthFlow(flows[flowType]);
        }
    }
    return result;
}
function normalizeOAuthFlow(flow) {
    return {
        authorizationUrl: optStr(flow.authorizationUrl),
        tokenUrl: optStr(flow.tokenUrl),
        refreshUrl: optStr(flow.refreshUrl),
        scopes: (flow.scopes ?? {}),
    };
}
// ── Webhooks ───────────────────────────────────────────────────────
function normalizeWebhooks(webhooks) {
    if (!webhooks)
        return [];
    return normalizeOperations(webhooks);
}
// ── Shared ─────────────────────────────────────────────────────────
function normalizeExternalDocs(docs) {
    if (!docs?.url)
        return undefined;
    return {
        url: str(docs.url, ""),
        description: optStr(docs.description),
    };
}
function normalizeExamples(examples) {
    const result = {};
    for (const [name, e] of Object.entries(examples)) {
        result[name] = {
            summary: optStr(e.summary),
            description: optStr(e.description),
            value: e.value,
            externalValue: optStr(e.externalValue),
        };
    }
    return result;
}
// ── Utility helpers ────────────────────────────────────────────────
function str(value, fallback) {
    return typeof value === "string" ? value : fallback;
}
function optStr(value) {
    return typeof value === "string" ? value : undefined;
}
